{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "worse-impact",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import gzip\n",
    "import sqlite3\n",
    "import zlib\n",
    "import io\n",
    "from sqlite3 import Error\n",
    "import pandas as pd\n",
    "from collections import OrderedDict\n",
    "\n",
    "from Bio import SeqIO \n",
    "from Bio.SeqRecord import SeqRecord\n",
    "from Bio.Seq import Seq"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "small-artist",
   "metadata": {},
   "source": [
    "# SQNce Mapping-related Tables"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "finished-threshold",
   "metadata": {},
   "outputs": [],
   "source": [
    "# TODO add documentation to all SQNce creation functions\n",
    "\n",
    "# Establish connection with SQNce.db, generating a new SQLite3 database if needed\n",
    "def sql_connection():\n",
    "    try:\n",
    "        con = sqlite3.connect('SQNce.db')\n",
    "        print(\"Connection established.\")\n",
    "        return(con)\n",
    "    except Error:\n",
    "        print(Error)\n",
    "\n",
    "# After establishing connection with SQNce create the specified tables\n",
    "def sql_table(con):\n",
    "    cursorObj = con.cursor()\n",
    "    \n",
    "    cursorObj.execute(\"\"\"CREATE TABLE IF NOT EXISTS mapping_traits(\n",
    "                         experiment text,\n",
    "                         trait text,\n",
    "                         description text,\n",
    "                         plot blob,\n",
    "                         processed BOOL)\n",
    "                         \"\"\")   \n",
    "\n",
    "    cursorObj.execute(\"\"\"CREATE TABLE IF NOT EXISTS mapping_results(\n",
    "                         experiment text,\n",
    "                         trait text,\n",
    "                         snp text,\n",
    "                         chrom text,\n",
    "                         pos integer,\n",
    "                         ref text,\n",
    "                         alt text,\n",
    "                         effect REAL,\n",
    "                         pval REAL)\n",
    "                         \"\"\")\n",
    "    cursorObj.execute(\"\"\"CREATE TABLE IF NOT EXISTS mapping_candidates(\n",
    "                         gene_id text,\n",
    "                         gene_symbol text,\n",
    "                         gene_description text, \n",
    "                         experiment text,\n",
    "                         trait text,\n",
    "                         candidate_count integer,\n",
    "                         snp text,\n",
    "                         chrom text,\n",
    "                         pos integer)\n",
    "                         \"\"\")   \n",
    "    con.commit()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "foreign-superintendent",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Current implementation requires re-parsing of all the input files to create SQNce\n",
    "# TODO SQNce update functions to parse input data only if not previously included \n",
    "#if os.path.exists(\"SQNce-proteomes.db\"): os.remove(\"SQNce-proteomes.db\")\n",
    "con = sql_connection()\n",
    "sql_table(con)\n",
    "con.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "identified-domain",
   "metadata": {},
   "source": [
    "# SQNce Mapping-related Functions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "accomplished-protest",
   "metadata": {},
   "outputs": [],
   "source": [
    "def mapping_traits_insert_traits(con, entities):\n",
    "    cursorObj = con.cursor()\n",
    "    cursorObj.executemany(\"\"\"INSERT INTO mapping_traits(\n",
    "                         experiment, \n",
    "                         trait, \n",
    "                         description,\n",
    "                         plot,\n",
    "                         processed) \n",
    "                         VALUES(?,?,?,?,?)\"\"\", entities)\n",
    "    con.commit()\n",
    "\n",
    "def mapping_traits_insert_results(con, entities):\n",
    "    cursorObj = con.cursor()\n",
    "    cursorObj.executemany(\"\"\"INSERT INTO mapping_results(\n",
    "                         experiment, \n",
    "                         trait, \n",
    "                         snp,\n",
    "                         chrom,\n",
    "                         pos,\n",
    "                         ref,\n",
    "                         alt,\n",
    "                         effect,\n",
    "                         pval) \n",
    "                         VALUES(?,?,?,?,?,?,?,?,?)\"\"\", entities)\n",
    "    con.commit()\n",
    "    \n",
    "def mapping_traits_insert_candidates(con, entities):\n",
    "    cursorObj = con.cursor()\n",
    "    cursorObj.execute(\"\"\"INSERT INTO mapping_candidates(\n",
    "                         experiment, \n",
    "                         trait, \n",
    "                         gene_id,\n",
    "                         gene_symbol,\n",
    "                         gene_description,\n",
    "                         candidate_count,\n",
    "                         snp,\n",
    "                         chrom,\n",
    "                         pos) \n",
    "                         VALUES(?,?,?,?,,?,?,?,?,?)\"\"\", entities)\n",
    "    con.commit()\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "chicken-locking",
   "metadata": {},
   "source": [
    "# SQNce Load the rMVP GLM GWAS Results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ordinary-frontier",
   "metadata": {},
   "outputs": [],
   "source": [
    "con = sql_connection()\n",
    "\n",
    "# The name of the experiment is also the name of the folder\n",
    "experiment = \"experiment\"\n",
    "\n",
    "fl_pvals_dict = {}\n",
    "fl_plots_dict = {}\n",
    "fl_name_set = set()\n",
    "for fl in os.listdir(experiment):\n",
    "    # Assume for now only my output files are used\n",
    "    if fl.split(\".\")[-1] == \"png\":\n",
    "        fl_name = fl.replace(\".png\", \"\")\n",
    "        fl_plots_dict[fl_name] = fl\n",
    "    if fl.split(\".\")[-1] == \"csv\":\n",
    "        fl_name = fl.replace(\".sig.csv\", \"\")\n",
    "        fl_pvals_dict[fl_name] = fl\n",
    "    fl_name_set.add(fl_name)\n",
    "fl_name_set\n",
    "\n",
    "\n",
    "all_traits_list = []\n",
    "all_results_list = []\n",
    "\n",
    "# The name of the trait is also the name of the file\n",
    "for trait in fl_name_set:\n",
    "    # Load the trait results dataframe\n",
    "    filename = os.path.join(experiment, fl_pvals_dict[trait])\n",
    "    df = pd.read_csv(filename)\n",
    "    if len(df) == 0:\n",
    "        continue\n",
    "    df = df.sort_values(df.columns[-1])\n",
    "\n",
    "    # Load the trait plot binary variable\n",
    "    filename = os.path.join(experiment, fl_plots_dict[trait])\n",
    "    with open(filename, 'rb') as file: plot_blob = file.read()\n",
    "    \n",
    "    all_traits_list.append([experiment, trait, \"rMVP GLM of 282_SC MR network with E5\", plot_blob, False])    \n",
    "\n",
    "    # List order: experiment,trait,snp,chrom,pos,ref,alt,effect,pval,plot\n",
    "    #selected_rows_ix = [0]\n",
    "    #selected_rows = df.iloc[selected_rows_ix,].values.tolist()\n",
    "    selected_rows = df.values.tolist()\n",
    "        \n",
    "    for row in selected_rows:\n",
    "        all_results_list.append([experiment, trait, row[0], row[1], row[2], row[3], row[4], row[5], row[7]])\n",
    "\n",
    "mapping_traits_insert_traits(con, all_traits_list)\n",
    "mapping_traits_insert_results(con, all_results_list)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "hundred-blair",
   "metadata": {},
   "outputs": [],
   "source": [
    "from io import BytesIO\n",
    "def pil_to_b64(im, enc_format=\"png\", **kwargs):\n",
    "    \"\"\"\n",
    "    Converts a PIL Image into base64 string for HTML displaying\n",
    "    :param im: PIL Image object\n",
    "    :param enc_format: The image format for displaying. If saved the image will have that extension.\n",
    "    :return: base64 encoding\n",
    "    \"\"\"\n",
    "\n",
    "    buff = BytesIO()\n",
    "    im.save(buff, format=enc_format, **kwargs)\n",
    "    encoded = base64.b64encode(buff.getvalue()).decode(\"utf-8\")\n",
    "\n",
    "    return encoded\n",
    "\n",
    "# https://stackoverflow.com/questions/14348442/how-do-i-display-a-pil-image-object-in-a-template#14348661\n",
    "import PIL\n",
    "import io\n",
    "from PIL import Image\n",
    "\n",
    "picture_stream = io.BytesIO(fig)\n",
    "picture = Image.open(picture_stream)\n",
    "# In dash you can send the image to html.Img like this:\n",
    "# html.Img(id=\"my-img\",className=\"image\", src=\"data:image/png;base64, \" + pil_to_b64(picture))\n",
    "\n",
    "# picture = PIL.Image.open(picture_stream)\n",
    "# picture.show()\n",
    "# pil_to_b64(picture)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
