{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "illegal-clock",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import gzip\n",
    "import sqlite3\n",
    "import zlib\n",
    "from sqlite3 import Error\n",
    "import pandas as pd\n",
    "from collections import OrderedDict\n",
    "\n",
    "from Bio import SeqIO \n",
    "from Bio.SeqRecord import SeqRecord\n",
    "from Bio.Seq import Seq"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "threaded-seller",
   "metadata": {},
   "source": [
    "# SQNce Creation Functions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "tested-yorkshire",
   "metadata": {},
   "outputs": [],
   "source": [
    "# TODO add documentation to all SQNce creation functions\n",
    "\n",
    "# Establish connection with SQNce.db, generating a new SQLite3 database if needed\n",
    "def sql_connection():\n",
    "    try:\n",
    "        con = sqlite3.connect('SQNce.db')\n",
    "        print(\"Connection established.\")\n",
    "        return(con)\n",
    "    except Error:\n",
    "        print(Error)\n",
    "\n",
    "# After establishing connection with SQNce create the specified tables\n",
    "def sql_table(con):\n",
    "    cursorObj = con.cursor()\n",
    "    cursorObj.execute(\"\"\"CREATE TABLE IF NOT EXISTS species(\n",
    "                         species_name text PRIMARY KEY, \n",
    "                         common_name text) \n",
    "                         WITHOUT ROWID\"\"\")\n",
    "    \n",
    "    cursorObj.execute(\"\"\"CREATE TABLE IF NOT EXISTS genotypes(\n",
    "                         genotype_id text PRIMARY KEY, \n",
    "                         species_name text, \n",
    "                         genotype_name text,\n",
    "                         FOREIGN KEY (species_name) REFERENCES species(species_name)) \n",
    "                         WITHOUT ROWID\"\"\")\n",
    "    \n",
    "    cursorObj.execute(\"\"\"CREATE TABLE IF NOT EXISTS gene_coordinates(\n",
    "                         gene_id text,\n",
    "                         genotype_id text,\n",
    "                         gene_chr text,\n",
    "                         gene_start integer,\n",
    "                         gene_end integer,\n",
    "                         gene_orientation text,\n",
    "                         FOREIGN KEY (genotype_id) REFERENCES genotypes(genotype_id)) \n",
    "                         \"\"\")\n",
    "\n",
    "    cursorObj.execute(\"\"\"CREATE TABLE IF NOT EXISTS protein_seqs(\n",
    "                         protein_isoform text,\n",
    "                         protein_id text,\n",
    "                         species_id text,\n",
    "                         genotype_id text,\n",
    "                         protein_length text,\n",
    "                         protein_sequence blob,\n",
    "                         FOREIGN KEY (genotype_id) REFERENCES genotypes(genotype_id)) \n",
    "                         \"\"\")\n",
    "    \n",
    "    cursorObj.execute(\"\"\"CREATE TABLE IF NOT EXISTS gene_families(\n",
    "                         protein_id text,\n",
    "                         species_id text,\n",
    "                         genotype_id text,\n",
    "                         source_id text,\n",
    "                         family_id text,\n",
    "                         family_name text,\n",
    "                         FOREIGN KEY (genotype_id) REFERENCES genotypes(genotype_id)) \n",
    "    \"\"\")\n",
    "    \n",
    "    cursorObj.execute(\"\"\"CREATE TABLE IF NOT EXISTS promoter_seqs(\n",
    "                         protein_id text,\n",
    "                         genotype_id text,\n",
    "                         promoter_kind text,\n",
    "                         promoter_length text,\n",
    "                         promoter_sequence blob,\n",
    "                         FOREIGN KEY (genotype_id) REFERENCES genotypes(genotype_id)) \n",
    "                         \"\"\")\n",
    "    \n",
    "    cursorObj.execute(\"\"\"CREATE TABLE IF NOT EXISTS gene_annotations(\n",
    "                         gene_id text PRIMARY KEY,\n",
    "                         gene_species text,\n",
    "                         gene_genotype text,\n",
    "                         annotation_source text,\n",
    "                         gene_annotation text,\n",
    "                         FOREIGN KEY (gene_genotype) REFERENCES genotypes(genotype_id)) \n",
    "                         WITHOUT ROWID\"\"\")\n",
    "    \n",
    "    cursorObj.execute(\"\"\"CREATE TABLE IF NOT EXISTS packages(\n",
    "                         rowid integer PRIMARY KEY,\n",
    "                         name text,\n",
    "                         version text,\n",
    "                         settings text) \n",
    "                         WITHOUT ROWID\"\"\")\n",
    "    \n",
    "    cursorObj.execute(\"\"\"CREATE TABLE IF NOT EXISTS studies(\n",
    "                         study_accession text PRIMARY KEY, \n",
    "                         tax_id integer,\n",
    "                         scientific_name text,\n",
    "                         instrument_model text,\n",
    "                         library_strategy text,\n",
    "                         description text)\n",
    "                         WITHOUT ROWID\"\"\")\n",
    "\n",
    "    cursorObj.execute(\"\"\"CREATE TABLE IF NOT EXISTS fastq(\n",
    "                         run_accession text PRIMARY KEY,\n",
    "                         study_accession text,\n",
    "                         read_count integer,\n",
    "                         sample_alias text,\n",
    "                         fastq_ftp text,\n",
    "                         fastq_md5 text,\n",
    "                         compression integer,\n",
    "                         FOREIGN KEY (study_accession) REFERENCES studies(study_accession)) \n",
    "                         WITHOUT ROWID\"\"\")\n",
    "    \n",
    "    cursorObj.execute(\"\"\"CREATE TABLE IF NOT EXISTS bam(\n",
    "                         run_accession text,\n",
    "                         study_accession text,\n",
    "                         sample_alias text,\n",
    "                         compression integer,\n",
    "                         filter integer,\n",
    "                         align integer,\n",
    "                         sort integer,\n",
    "                         FOREIGN KEY (run_accession) REFERENCES fastq(run_accession)) \n",
    "                         \"\"\")\n",
    "\n",
    "    con.commit()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "indian-tyler",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Connection established.\n"
     ]
    }
   ],
   "source": [
    "# Current implementation requires re-parsing of all the input files to create SQNce\n",
    "# TODO SQNce update functions to parse input data only if not previously included \n",
    "if os.path.exists(\"SQNce.db\"): os.remove(\"SQNce.db\")\n",
    "con = sql_connection()\n",
    "sql_table(con)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "piano-secondary",
   "metadata": {},
   "source": [
    "# SQNce Data Input Functions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "aquatic-identification",
   "metadata": {},
   "outputs": [],
   "source": [
    "# TODO add documentation to all SQNce Data Input Functions\n",
    "\n",
    "def species_insert(con, entities):\n",
    "    cursorObj = con.cursor()\n",
    "    cursorObj.execute(\"\"\"INSERT INTO species(\n",
    "                         species_name, common_name) \n",
    "                         VALUES(?, ?)\"\"\", entities)\n",
    "    con.commit()\n",
    "    \n",
    "def genotype_insert(con, entities):\n",
    "    cursorObj = con.cursor()\n",
    "    cursorObj.execute(\"\"\"INSERT INTO genotypes(\n",
    "                         genotype_id, \n",
    "                         species_name, \n",
    "                         genotype_name) \n",
    "                         VALUES(?, ?, ?)\"\"\", entities)\n",
    "    con.commit()\n",
    "\n",
    "# https://stackoverflow.com/questions/18219779/bulk-insert-huge-data-into-sqlite-using-python\n",
    "def gene_coordinates_insert(con, entity_list):\n",
    "    cursorObj = con.cursor()\n",
    "    cursorObj.executemany(\"\"\"INSERT INTO gene_coordinates(\n",
    "                         gene_id,\n",
    "                         genotype_id,\n",
    "                         gene_chr, \n",
    "                         gene_start,  \n",
    "                         gene_end,\n",
    "                         gene_orientation) \n",
    "                         VALUES(?, ?, ?, ?, ?, ?)\"\"\", entity_list)\n",
    "    con.commit()\n",
    "\n",
    "    \n",
    "    \n",
    "# https://stackoverflow.com/questions/18219779/bulk-insert-huge-data-into-sqlite-using-python\n",
    "def protein_seq_insert(con, entity_list):\n",
    "    cursorObj = con.cursor()\n",
    "    cursorObj.executemany(\"\"\"INSERT INTO protein_seqs(\n",
    "                         protein_isoform,\n",
    "                         protein_id,\n",
    "                         species_id,\n",
    "                         genotype_id, \n",
    "                         protein_length,  \n",
    "                         protein_sequence) \n",
    "                         VALUES(?, ?, ?, ?, ?, ?)\"\"\", entity_list)\n",
    "    con.commit()\n",
    "    \n",
    "def family_annotation_insert(con, entity_list):\n",
    "    cursorObj = con.cursor()\n",
    "    cursorObj.executemany(\"\"\"INSERT INTO gene_families(\n",
    "                         protein_id,\n",
    "                         species_id,\n",
    "                         genotype_id,\n",
    "                         source_id,\n",
    "                         family_id,  \n",
    "                         family_name) \n",
    "                         VALUES(?, ?, ?, ?, ?, ?)\"\"\", entity_list)\n",
    "    con.commit()\n",
    "    \n",
    "# https://stackoverflow.com/questions/18219779/bulk-insert-huge-data-into-sqlite-using-python\n",
    "def promoter_seq_insert(con, entity_list):\n",
    "    cursorObj = con.cursor()\n",
    "    cursorObj.executemany(\"\"\"INSERT INTO promoter_seqs(\n",
    "                         protein_id,\n",
    "                         genotype_id,\n",
    "                         promoter_kind,\n",
    "                         promoter_length,  \n",
    "                         promoter_sequence) \n",
    "                         VALUES(?, ?, ?, ?, ?)\"\"\", entity_list)\n",
    "    con.commit()\n",
    "\n",
    "    \n",
    "def gene_annotation_insert(con, entity_list):\n",
    "    cursorObj = con.cursor()\n",
    "    cursorObj.executemany(\"\"\"INSERT INTO gene_annotations(\n",
    "                         gene_id,\n",
    "                         gene_species,\n",
    "                         gene_genotype,\n",
    "                         annotation_source,\n",
    "                         gene_annotation) \n",
    "                         VALUES(?, ?, ?, ?, ?)\"\"\", entity_list)\n",
    "    con.commit()\n",
    "\n",
    "def packages_insert(con, entities):\n",
    "    cursorObj = con.cursor()\n",
    "    cursorObj.execute(\"\"\"INSERT INTO packages(\n",
    "                         rowid, \n",
    "                         name,\n",
    "                         version,\n",
    "                         settings) \n",
    "                         VALUES(?, ?, ?, ?)\"\"\", entities)\n",
    "    con.commit()\n",
    "    \n",
    "def studies_insert(con, entities):\n",
    "    cursorObj = con.cursor()\n",
    "    cursorObj.execute(\"\"\"INSERT INTO studies(\n",
    "                         study_accession, \n",
    "                         tax_id, \n",
    "                         scientific_name,\n",
    "                         instrument_model,\n",
    "                         library_strategy,\n",
    "                         description) \n",
    "                         VALUES(?, ?, ?, ?, ?, ?)\"\"\", entities)\n",
    "    con.commit()\n",
    "\n",
    "# https://stackoverflow.com/questions/18219779/bulk-insert-huge-data-into-sqlite-using-python\n",
    "def fastq_insert(con, entities):\n",
    "    cursorObj = con.cursor()\n",
    "    cursorObj.execute(\"\"\"INSERT INTO fastq(\n",
    "                         run_accession, \n",
    "                         study_accession, \n",
    "                         read_count, \n",
    "                         sample_alias,  \n",
    "                         fastq_ftp,\n",
    "                         fastq_md5,\n",
    "                         compression) \n",
    "                         VALUES(?, ?, ?, ?, ?, ?, ?)\"\"\", entities)\n",
    "    con.commit()\n",
    "    \n",
    "def bam_insert(con, entities):\n",
    "    cursorObj = con.cursor()\n",
    "    cursorObj.execute(\"\"\"INSERT INTO bam(\n",
    "                         run_accession,\n",
    "                         study_accession,\n",
    "                         sample_alias,\n",
    "                         compression,\n",
    "                         filter,\n",
    "                         align,\n",
    "                         sort) \n",
    "                         VALUES(?, ?, ?, ?, ?, ?, ?)\"\"\", entities)\n",
    "    con.commit()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "democratic-spectacular",
   "metadata": {},
   "outputs": [],
   "source": [
    "# SQNce database is initiated using predifined TSV files\n",
    "# TSV either contain the input data or reference input files to parse\n",
    "# TODO add documentation to SQNce data input parsing\n",
    "\n",
    "df = pd.read_csv(\"init/species.tsv\", sep=\"\\t\")\n",
    "for index, row in df.iterrows():\n",
    "    species_insert(con, entities=list(row))\n",
    "\n",
    "df = pd.read_csv(\"init/genotypes.tsv\", sep=\"\\t\")\n",
    "for index, row in df.iterrows():\n",
    "    genotype_insert(con, entities=list(row))\n",
    "\n",
    "#######################################################################\n",
    "############################ Coordinates ##############################\n",
    "#######################################################################\n",
    "df = pd.read_csv(\"init/coordinates.tsv\", sep=\"\\t\")\n",
    "for index, row in df.iterrows():\n",
    "    gff3_file = gzip.open(\"inputs/gff3/\"+row[3], mode='rt')\n",
    "    gene_coordinate_list = []\n",
    "    for gene in gff3_file:\n",
    "        gene = gene.split(\"\\t\")\n",
    "        if len(gene) == 1:\n",
    "            # skip the first row\n",
    "            continue\n",
    "        if gene[2] == \"gene\":\n",
    "            gene_id = gene[-1].split(\";\")\n",
    "            gene_id = [i for i in gene_id if i.startswith('Name=')][0].replace('Name=', '').replace('\\n', '') # delete \\n if exists\n",
    "            # Append list of: gene ID, genotype, chromsome, start, end, orientation\n",
    "            gene_coordinate_list.append([gene_id, row[1], gene[0], gene[3], gene[4], gene[6]])\n",
    "        else:\n",
    "            continue\n",
    "    gene_coordinates_insert(con, gene_coordinate_list)    \n",
    "\n",
    "# Create a secondary key on the name column\n",
    "cursorObj = con.cursor()\n",
    "cursorObj.execute(\"CREATE INDEX coordinate_index_start ON gene_coordinates(genotype_id, gene_chr, gene_start)\")\n",
    "cursorObj.execute(\"CREATE INDEX coordinate_index_end ON gene_coordinates(genotype_id, gene_chr, gene_end)\")\n",
    "con.commit()\n",
    "\n",
    "#######################################################################\n",
    "########################### Protein Seqs ##############################\n",
    "#######################################################################\n",
    "df = pd.read_csv(\"init/proteins.tsv\", sep=\"\\t\")\n",
    "for index, row in df.iterrows():\n",
    "    fasta_file = gzip.open(\"inputs/proteins/\"+row[2], mode='rt')\n",
    "    protein_seq_list = []\n",
    "    for seq in SeqIO.parse(fasta_file, \"fasta\"):\n",
    "        # Extract the locus ID that doesn't contain the isoform\n",
    "        locus = seq.description.split(\" \")\n",
    "        for ls in locus:\n",
    "            if \"locus\" in ls:\n",
    "                locus = ls[6:]\n",
    "        # Protein sequences are saved as as a Binary data type for compression\n",
    "        protein_seq_list.append([seq.id, locus, row[0], row[1], len(seq.seq), \n",
    "                                 sqlite3.Binary(zlib.compress(str(seq.seq).encode('utf-8')))])\n",
    "    protein_seq_insert(con, protein_seq_list)\n",
    "# Create keys for protein IDs with isoform and without\n",
    "cursorObj = con.cursor()\n",
    "cursorObj.execute(\"CREATE INDEX proteins_isoforms_index ON protein_seqs(protein_isoform)\")\n",
    "cursorObj.execute(\"CREATE INDEX proteins_ids_index ON protein_seqs(protein_id)\")\n",
    "con.commit()    \n",
    "\n",
    "#######################################################################\n",
    "########################### Family Annotations ##############################\n",
    "#######################################################################\n",
    "df = pd.read_csv(\"init/families.tsv\", sep=\"\\t\")\n",
    "for index, row in df.iterrows():\n",
    "    annot = pd.read_csv(\"inputs/families/\"+row[3], sep=\"\\t\")\n",
    "    annot.insert(1, 'species', row[0])\n",
    "    annot.insert(2, 'genotype', row[1])\n",
    "    annot.insert(3, 'source', row[2])\n",
    "    family_annotation_insert(con, annot.values.tolist())\n",
    "# Create keys for protein IDs with isoform and without\n",
    "cursorObj = con.cursor()\n",
    "cursorObj.execute(\"CREATE INDEX family_gene_index ON gene_families(protein_id)\")\n",
    "#cursorObj.execute(\"CREATE INDEX family_species_index ON gene_families(species_id)\")\n",
    "#cursorObj.execute(\"CREATE INDEX family_genotype_index ON gene_families(genotype_id)\")\n",
    "#cursorObj.execute(\"CREATE INDEX family_id_index ON gene_families(family_id)\")\n",
    "cursorObj.execute(\"CREATE INDEX family_name_index ON gene_families(family_name)\")\n",
    "cursorObj.execute(\"CREATE INDEX family_genotype_name_index ON gene_families(genotype_id, family_name)\")\n",
    "con.commit()    \n",
    "\n",
    "\n",
    "\n",
    "#######################################################################\n",
    "########################### Promoter Seqs #############################\n",
    "#######################################################################\n",
    "df = pd.read_csv(\"init/promoters.tsv\", sep=\"\\t\")\n",
    "for index, row in df.iterrows():\n",
    "    fasta_file = gzip.open(\"inputs/promoters/\"+row[5]+\"_1kb_ATG.fasta.gz\", mode='rt')\n",
    "    promoter_seq_list = []\n",
    "    for seq in SeqIO.parse(fasta_file, \"fasta\"):\n",
    "        # TODO change db to number key to avoid duplicated name problems\n",
    "        if row[0] == \"ZmB73v3\": \n",
    "            seq.id = seq.id + \"v3\"\n",
    "        # Protein sequences are saved as as a Binary data type for compression\n",
    "        promoter_seq_list.append([seq.id, row[0], \"ATG\", row[4], \n",
    "                                 sqlite3.Binary(zlib.compress(str(seq.seq).encode('utf-8')))])\n",
    "    promoter_seq_insert(con, promoter_seq_list)\n",
    "    \n",
    "    fasta_file = gzip.open(\"inputs/promoters/\"+row[5]+\"_1kb_TSS.fasta.gz\", mode='rt')\n",
    "    promoter_seq_list = []\n",
    "    for seq in SeqIO.parse(fasta_file, \"fasta\"):\n",
    "        # TODO change db to number key to avoid duplicated name problems\n",
    "        if row[0] == \"ZmB73v3\": \n",
    "            seq.id = seq.id + \"v3\"\n",
    "        # Protein sequences are saved as as a Binary data type for compression\n",
    "        promoter_seq_list.append([seq.id, row[0], \"TSS\", row[4], \n",
    "                                 sqlite3.Binary(zlib.compress(str(seq.seq).encode('utf-8')))])\n",
    "    promoter_seq_insert(con, promoter_seq_list)\n",
    "    \n",
    "# Create a secondary key on the name column\n",
    "cursorObj = con.cursor()\n",
    "cursorObj.execute(\"CREATE INDEX promoter_index ON promoter_seqs(protein_id, promoter_kind)\")\n",
    "con.commit()\n",
    "\n",
    "#######################################################################\n",
    "############################ Annotations ##############################\n",
    "#######################################################################\n",
    "\n",
    "df = pd.read_csv(\"init/annotation_list.tsv\", sep=\"\\t\")\n",
    "# Every element is: gene_id, gene_species, gene_genotype, gene_annotation, annotation_source\n",
    "gene_annotation_list = []\n",
    "for index, row in df.iterrows():\n",
    "    annot = pd.read_csv(\"inputs/annotations/\"+row[3], sep=\"\\t\")\n",
    "    annot = annot.drop_duplicates(subset=\"locusName\")\n",
    "\n",
    "    if row[0]==\"Arabidopsis thaliana\":\n",
    "        annot = annot[[\"locusName\", \"rice-defline\"]]\n",
    "    elif row[0] in [\"Panicum hallii\", \"Pharus latifolius\", \"Solanum lycopersicum\", \"Vigna unguiculata\"]:\n",
    "        annot = annot[[\"locusName\", \"Best-hit-arabi-defline\"]]        \n",
    "    else:\n",
    "        annot = annot[[\"locusName\", \"arabi-defline\"]]\n",
    "    annot.insert(1, 'species', row[0])\n",
    "    annot.insert(2, 'genotype', row[1])\n",
    "    annot.insert(3, 'source', row[2])\n",
    "    gene_annotation_insert(con, annot.values.tolist())\n",
    "\n",
    "#######################################################################\n",
    "########################### RNA-seq Files #############################\n",
    "#######################################################################\n",
    "    \n",
    "\n",
    "df = pd.read_csv(\"inputs/omics/packages.tsv\", sep=\"\\t\")\n",
    "for index, row in df.iterrows():\n",
    "    packages_insert(con, entities=list(row))\n",
    "\n",
    "df = pd.read_csv(\"inputs/omics/studies.tsv\", sep=\"\\t\")\n",
    "for index, row in df.iterrows():\n",
    "    studies_insert(con, entities=list(row))\n",
    "\n",
    "df = pd.read_csv(\"inputs/omics/fastq.tsv\", sep=\"\\t\")\n",
    "for index, row in df.iterrows():\n",
    "    fastq_insert(con, entities=list(row))\n",
    "\n",
    "df = pd.read_csv(\"inputs/omics/bam.tsv\", sep=\"\\t\")\n",
    "for index, row in df.iterrows():\n",
    "    bam_insert(con, entities=list(row))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "oriental-globe",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene</th>\n",
       "      <th>species</th>\n",
       "      <th>genotype</th>\n",
       "      <th>source</th>\n",
       "      <th>SSF</th>\n",
       "      <th>eval</th>\n",
       "      <th>description</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Zm00001d020134</td>\n",
       "      <td>Zea mays</td>\n",
       "      <td>B73</td>\n",
       "      <td>SUPERFAMILY</td>\n",
       "      <td>37958</td>\n",
       "      <td>0.0</td>\n",
       "      <td>P-loop containing nucleoside triphosphate hydr...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Zm00001d020134</td>\n",
       "      <td>Zea mays</td>\n",
       "      <td>B73</td>\n",
       "      <td>SUPERFAMILY</td>\n",
       "      <td>37958</td>\n",
       "      <td>0.0</td>\n",
       "      <td>P-loop containing nucleoside triphosphate hydr...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Zm00001d021647</td>\n",
       "      <td>Zea mays</td>\n",
       "      <td>B73</td>\n",
       "      <td>SUPERFAMILY</td>\n",
       "      <td>37958</td>\n",
       "      <td>0.0</td>\n",
       "      <td>P-loop containing nucleoside triphosphate hydr...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Zm00001d021647</td>\n",
       "      <td>Zea mays</td>\n",
       "      <td>B73</td>\n",
       "      <td>SUPERFAMILY</td>\n",
       "      <td>37958</td>\n",
       "      <td>0.0</td>\n",
       "      <td>P-loop containing nucleoside triphosphate hydr...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Zm00001d020208</td>\n",
       "      <td>Zea mays</td>\n",
       "      <td>B73</td>\n",
       "      <td>SUPERFAMILY</td>\n",
       "      <td>37958</td>\n",
       "      <td>0.0</td>\n",
       "      <td>P-loop containing nucleoside triphosphate hydr...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36867</th>\n",
       "      <td>Zm00001d013760</td>\n",
       "      <td>Zea mays</td>\n",
       "      <td>B73</td>\n",
       "      <td>SUPERFAMILY</td>\n",
       "      <td>46406</td>\n",
       "      <td>0.0</td>\n",
       "      <td>ENTH/VHS domain</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36868</th>\n",
       "      <td>Zm00001d017951</td>\n",
       "      <td>Zea mays</td>\n",
       "      <td>B73</td>\n",
       "      <td>SUPERFAMILY</td>\n",
       "      <td>49418</td>\n",
       "      <td>0.0</td>\n",
       "      <td>Nudix</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36869</th>\n",
       "      <td>Zm00001d012826</td>\n",
       "      <td>Zea mays</td>\n",
       "      <td>B73</td>\n",
       "      <td>SUPERFAMILY</td>\n",
       "      <td>54271</td>\n",
       "      <td>0.0</td>\n",
       "      <td>Ribonuclease PH domain 2-like</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36870</th>\n",
       "      <td>Zm00001d013151</td>\n",
       "      <td>Zea mays</td>\n",
       "      <td>B73</td>\n",
       "      <td>SUPERFAMILY</td>\n",
       "      <td>51412</td>\n",
       "      <td>0.0</td>\n",
       "      <td>Lipocalins</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36871</th>\n",
       "      <td>Zm00001d016571</td>\n",
       "      <td>Zea mays</td>\n",
       "      <td>B73</td>\n",
       "      <td>SUPERFAMILY</td>\n",
       "      <td>35325</td>\n",
       "      <td>0.0</td>\n",
       "      <td>Sm-like ribonucleoproteins</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>36872 rows × 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                 gene   species genotype       source    SSF  eval  \\\n",
       "0      Zm00001d020134  Zea mays      B73  SUPERFAMILY  37958   0.0   \n",
       "1      Zm00001d020134  Zea mays      B73  SUPERFAMILY  37958   0.0   \n",
       "2      Zm00001d021647  Zea mays      B73  SUPERFAMILY  37958   0.0   \n",
       "3      Zm00001d021647  Zea mays      B73  SUPERFAMILY  37958   0.0   \n",
       "4      Zm00001d020208  Zea mays      B73  SUPERFAMILY  37958   0.0   \n",
       "...               ...       ...      ...          ...    ...   ...   \n",
       "36867  Zm00001d013760  Zea mays      B73  SUPERFAMILY  46406   0.0   \n",
       "36868  Zm00001d017951  Zea mays      B73  SUPERFAMILY  49418   0.0   \n",
       "36869  Zm00001d012826  Zea mays      B73  SUPERFAMILY  54271   0.0   \n",
       "36870  Zm00001d013151  Zea mays      B73  SUPERFAMILY  51412   0.0   \n",
       "36871  Zm00001d016571  Zea mays      B73  SUPERFAMILY  35325   0.0   \n",
       "\n",
       "                                             description  \n",
       "0      P-loop containing nucleoside triphosphate hydr...  \n",
       "1      P-loop containing nucleoside triphosphate hydr...  \n",
       "2      P-loop containing nucleoside triphosphate hydr...  \n",
       "3      P-loop containing nucleoside triphosphate hydr...  \n",
       "4      P-loop containing nucleoside triphosphate hydr...  \n",
       "...                                                  ...  \n",
       "36867                                    ENTH/VHS domain  \n",
       "36868                                              Nudix  \n",
       "36869                      Ribonuclease PH domain 2-like  \n",
       "36870                                         Lipocalins  \n",
       "36871                         Sm-like ribonucleoproteins  \n",
       "\n",
       "[36872 rows x 7 columns]"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "annot"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "developed-colon",
   "metadata": {},
   "source": [
    "# SQNce Query Functions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "acquired-rally",
   "metadata": {},
   "outputs": [
    {
     "ename": "SyntaxError",
     "evalue": "'return' outside function (<ipython-input-27-5da10337c74d>, line 7)",
     "output_type": "error",
     "traceback": [
      "\u001b[1;36m  File \u001b[1;32m\"<ipython-input-27-5da10337c74d>\"\u001b[1;36m, line \u001b[1;32m7\u001b[0m\n\u001b[1;33m    return(html.P(\"Number of columns is not 2. Use tab-seperated values.\"))\u001b[0m\n\u001b[1;37m    ^\u001b[0m\n\u001b[1;31mSyntaxError\u001b[0m\u001b[1;31m:\u001b[0m 'return' outside function\n"
     ]
    }
   ],
   "source": [
    "rows = \"\"\"Chr2\\t12345\\nChr3\\t54354\\nChr2\\t5234354\\n\"\"\".split(\"\\n\")\n",
    "coordinate_list = [row.split(\"\\t\") for row in rows]\n",
    "if coordinate_list[-1]==[\"\"]:\n",
    "    coordinate_list = coordinate_list[:-1]\n",
    "for row in coordinate_list:\n",
    "    if len(row) != 2:\n",
    "        return(html.P(\"Number of columns is not 2. Use tab-seperated values.\"))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "id": "permanent-deviation",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Query to find neighboring genes\n",
    "def get_SNP_neighbors(genotype, chromsome, coordinate, distance):\n",
    "    con = sqlite3.connect('SQNce.db')\n",
    "    cursorObj = con.cursor()\n",
    "    df = pd.read_sql_query('''SELECT * \n",
    "                     FROM gene_coordinates \n",
    "                     WHERE genotype_id = \"{0}\"\n",
    "                     AND gene_chr = \"{1}\"\n",
    "                     AND gene_start BETWEEN {2} AND {3}\n",
    "                     \n",
    "                     UNION ALL\n",
    "                     \n",
    "                     SELECT * \n",
    "                     FROM gene_coordinates \n",
    "                     WHERE genotype_id = \"{0}\"\n",
    "                     AND gene_chr = \"{1}\"\n",
    "                     AND gene_start BETWEEN {2} AND {3}\n",
    "                     '''.format(genotype, chromsome, coordinate-distance, coordinate+distance), con)\n",
    "    # Should check why it returns the same row twice, probably need to correct the query\n",
    "    df = df.drop_duplicates()\n",
    "    df.insert(0, 'Query', pd.Series([\"_\".join([chromsome, str(coordinate)]) for x in range(len(df.index))]))\n",
    "    return(df)\n",
    "\n",
    "df = pd.DataFrame()\n",
    "for entity in [[\"Chr2\", 19681637], [\"Chr2\", 1234564], [\"Chr4\", 1234564], [\"Chr2\", 19681638]]:\n",
    "      df = pd.concat([df, get_SNP_neighbors(\"Arabidopsis\", entity[0], entity[1], 10000)])  \n",
    "#df.drop_duplicates(subset=[\"gene_id\"]).reset_index()\n",
    "#df1 = get_SNP_neighbors(\"Arabidopsis\", \"Chr2\", 19681637, 10000)\n",
    "#df1\n",
    "#pd.concat([df1, df1])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "id": "metallic-keyboard",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Query</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>genotype_id</th>\n",
       "      <th>gene_chr</th>\n",
       "      <th>gene_start</th>\n",
       "      <th>gene_end</th>\n",
       "      <th>gene_orientation</th>\n",
       "      <th>annotation</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Chr2_19681637</td>\n",
       "      <td>AT2G48110</td>\n",
       "      <td>Arabidopsis</td>\n",
       "      <td>Chr2</td>\n",
       "      <td>19673293</td>\n",
       "      <td>19679711</td>\n",
       "      <td>+</td>\n",
       "      <td>structural constituent of ribosome, putative, ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Chr2_19681637</td>\n",
       "      <td>AT2G48120</td>\n",
       "      <td>Arabidopsis</td>\n",
       "      <td>Chr2</td>\n",
       "      <td>19679730</td>\n",
       "      <td>19681546</td>\n",
       "      <td>+</td>\n",
       "      <td>PAC, putative, expressed</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Chr2_19681637</td>\n",
       "      <td>AT2G48121</td>\n",
       "      <td>Arabidopsis</td>\n",
       "      <td>Chr2</td>\n",
       "      <td>19681637</td>\n",
       "      <td>19682391</td>\n",
       "      <td>+</td>\n",
       "      <td>chloroplast ribonuclease III domain protein, p...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Chr2_19681637</td>\n",
       "      <td>AT2G48130</td>\n",
       "      <td>Arabidopsis</td>\n",
       "      <td>Chr2</td>\n",
       "      <td>19685066</td>\n",
       "      <td>19685993</td>\n",
       "      <td>-</td>\n",
       "      <td>LTPL78 - Protease inhibitor/seed storage/LTP f...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Chr2_19681637</td>\n",
       "      <td>AT2G48140</td>\n",
       "      <td>Arabidopsis</td>\n",
       "      <td>Chr2</td>\n",
       "      <td>19686409</td>\n",
       "      <td>19687664</td>\n",
       "      <td>+</td>\n",
       "      <td>LTPL99 - Protease inhibitor/seed storage/LTP f...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>Chr2_19681637</td>\n",
       "      <td>AT2G48150</td>\n",
       "      <td>Arabidopsis</td>\n",
       "      <td>Chr2</td>\n",
       "      <td>19687962</td>\n",
       "      <td>19689173</td>\n",
       "      <td>-</td>\n",
       "      <td>glutathione peroxidase domain containing prote...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>Chr2_19681637</td>\n",
       "      <td>AT2G48160</td>\n",
       "      <td>Arabidopsis</td>\n",
       "      <td>Chr2</td>\n",
       "      <td>19689409</td>\n",
       "      <td>19696821</td>\n",
       "      <td>-</td>\n",
       "      <td>PWWP domain containing protein, expressed</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Chr2_1234564</td>\n",
       "      <td>AT2G03955</td>\n",
       "      <td>Arabidopsis</td>\n",
       "      <td>Chr2</td>\n",
       "      <td>1238677</td>\n",
       "      <td>1239071</td>\n",
       "      <td>+</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Chr4_1234564</td>\n",
       "      <td>AT4G02770</td>\n",
       "      <td>Arabidopsis</td>\n",
       "      <td>Chr4</td>\n",
       "      <td>1229111</td>\n",
       "      <td>1229945</td>\n",
       "      <td>-</td>\n",
       "      <td>photosystem I reaction center subunit II, chlo...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Chr4_1234564</td>\n",
       "      <td>AT4G02780</td>\n",
       "      <td>Arabidopsis</td>\n",
       "      <td>Chr4</td>\n",
       "      <td>1237767</td>\n",
       "      <td>1244813</td>\n",
       "      <td>-</td>\n",
       "      <td>ent-kaurene synthase, chloroplast precursor, p...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Chr2_19681638</td>\n",
       "      <td>AT2G48110</td>\n",
       "      <td>Arabidopsis</td>\n",
       "      <td>Chr2</td>\n",
       "      <td>19673293</td>\n",
       "      <td>19679711</td>\n",
       "      <td>+</td>\n",
       "      <td>structural constituent of ribosome, putative, ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Chr2_19681638</td>\n",
       "      <td>AT2G48120</td>\n",
       "      <td>Arabidopsis</td>\n",
       "      <td>Chr2</td>\n",
       "      <td>19679730</td>\n",
       "      <td>19681546</td>\n",
       "      <td>+</td>\n",
       "      <td>PAC, putative, expressed</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Chr2_19681638</td>\n",
       "      <td>AT2G48121</td>\n",
       "      <td>Arabidopsis</td>\n",
       "      <td>Chr2</td>\n",
       "      <td>19681637</td>\n",
       "      <td>19682391</td>\n",
       "      <td>+</td>\n",
       "      <td>chloroplast ribonuclease III domain protein, p...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Chr2_19681638</td>\n",
       "      <td>AT2G48130</td>\n",
       "      <td>Arabidopsis</td>\n",
       "      <td>Chr2</td>\n",
       "      <td>19685066</td>\n",
       "      <td>19685993</td>\n",
       "      <td>-</td>\n",
       "      <td>LTPL78 - Protease inhibitor/seed storage/LTP f...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Chr2_19681638</td>\n",
       "      <td>AT2G48140</td>\n",
       "      <td>Arabidopsis</td>\n",
       "      <td>Chr2</td>\n",
       "      <td>19686409</td>\n",
       "      <td>19687664</td>\n",
       "      <td>+</td>\n",
       "      <td>LTPL99 - Protease inhibitor/seed storage/LTP f...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>Chr2_19681638</td>\n",
       "      <td>AT2G48150</td>\n",
       "      <td>Arabidopsis</td>\n",
       "      <td>Chr2</td>\n",
       "      <td>19687962</td>\n",
       "      <td>19689173</td>\n",
       "      <td>-</td>\n",
       "      <td>glutathione peroxidase domain containing prote...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>Chr2_19681638</td>\n",
       "      <td>AT2G48160</td>\n",
       "      <td>Arabidopsis</td>\n",
       "      <td>Chr2</td>\n",
       "      <td>19689409</td>\n",
       "      <td>19696821</td>\n",
       "      <td>-</td>\n",
       "      <td>PWWP domain containing protein, expressed</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           Query    gene_id  genotype_id gene_chr  gene_start  gene_end  \\\n",
       "0  Chr2_19681637  AT2G48110  Arabidopsis     Chr2    19673293  19679711   \n",
       "1  Chr2_19681637  AT2G48120  Arabidopsis     Chr2    19679730  19681546   \n",
       "2  Chr2_19681637  AT2G48121  Arabidopsis     Chr2    19681637  19682391   \n",
       "3  Chr2_19681637  AT2G48130  Arabidopsis     Chr2    19685066  19685993   \n",
       "4  Chr2_19681637  AT2G48140  Arabidopsis     Chr2    19686409  19687664   \n",
       "5  Chr2_19681637  AT2G48150  Arabidopsis     Chr2    19687962  19689173   \n",
       "6  Chr2_19681637  AT2G48160  Arabidopsis     Chr2    19689409  19696821   \n",
       "0   Chr2_1234564  AT2G03955  Arabidopsis     Chr2     1238677   1239071   \n",
       "0   Chr4_1234564  AT4G02770  Arabidopsis     Chr4     1229111   1229945   \n",
       "1   Chr4_1234564  AT4G02780  Arabidopsis     Chr4     1237767   1244813   \n",
       "0  Chr2_19681638  AT2G48110  Arabidopsis     Chr2    19673293  19679711   \n",
       "1  Chr2_19681638  AT2G48120  Arabidopsis     Chr2    19679730  19681546   \n",
       "2  Chr2_19681638  AT2G48121  Arabidopsis     Chr2    19681637  19682391   \n",
       "3  Chr2_19681638  AT2G48130  Arabidopsis     Chr2    19685066  19685993   \n",
       "4  Chr2_19681638  AT2G48140  Arabidopsis     Chr2    19686409  19687664   \n",
       "5  Chr2_19681638  AT2G48150  Arabidopsis     Chr2    19687962  19689173   \n",
       "6  Chr2_19681638  AT2G48160  Arabidopsis     Chr2    19689409  19696821   \n",
       "\n",
       "  gene_orientation                                         annotation  \n",
       "0                +  structural constituent of ribosome, putative, ...  \n",
       "1                +                           PAC, putative, expressed  \n",
       "2                +  chloroplast ribonuclease III domain protein, p...  \n",
       "3                -  LTPL78 - Protease inhibitor/seed storage/LTP f...  \n",
       "4                +  LTPL99 - Protease inhibitor/seed storage/LTP f...  \n",
       "5                -  glutathione peroxidase domain containing prote...  \n",
       "6                -          PWWP domain containing protein, expressed  \n",
       "0                +                                               None  \n",
       "0                -  photosystem I reaction center subunit II, chlo...  \n",
       "1                -  ent-kaurene synthase, chloroplast precursor, p...  \n",
       "0                +  structural constituent of ribosome, putative, ...  \n",
       "1                +                           PAC, putative, expressed  \n",
       "2                +  chloroplast ribonuclease III domain protein, p...  \n",
       "3                -  LTPL78 - Protease inhibitor/seed storage/LTP f...  \n",
       "4                +  LTPL99 - Protease inhibitor/seed storage/LTP f...  \n",
       "5                -  glutathione peroxidase domain containing prote...  \n",
       "6                -          PWWP domain containing protein, expressed  "
      ]
     },
     "execution_count": 39,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Query to find neighboring genes and annotate them\n",
    "\n",
    "df[\"annotation\"] = annotation_select(con, df[\"gene_id\"].to_list())\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "smooth-developer",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>species_name</th>\n",
       "      <th>common_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Arabidopsis halleri</td>\n",
       "      <td>Ahalleri</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Arabidopsis lyrata</td>\n",
       "      <td>Alyrata</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Arabidopsis thaliana</td>\n",
       "      <td>Arabidopsis</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Brassica oleracea</td>\n",
       "      <td>Boleracea</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Brassica rapa</td>\n",
       "      <td>Brapa</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>Glycine max</td>\n",
       "      <td>Soybean</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>Hordeum vulgare</td>\n",
       "      <td>Barley</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>Medicago truncatula</td>\n",
       "      <td>Barrel medic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>Oropetium thomaeum</td>\n",
       "      <td>Resurrection grass</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>Oryza sativa</td>\n",
       "      <td>Rice</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>Oryza sativa Kitaake</td>\n",
       "      <td>Okitaake</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>Panicum hallii</td>\n",
       "      <td>Hall's panicgrass</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>Pharus latifolius</td>\n",
       "      <td>Pharus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>Phaseolus vulgaris</td>\n",
       "      <td>Common bean</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>Solanum lycopersicum</td>\n",
       "      <td>Tomato</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>Sorghum bicolor</td>\n",
       "      <td>Sorghum</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>Thinopyrum intermedium</td>\n",
       "      <td>Wheatgrass</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>Triticum aestivum</td>\n",
       "      <td>Wheat</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>Vigna unguiculata</td>\n",
       "      <td>Cowpea</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>Zea mays</td>\n",
       "      <td>Maize</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "              species_name         common_name\n",
       "0      Arabidopsis halleri            Ahalleri\n",
       "1       Arabidopsis lyrata             Alyrata\n",
       "2     Arabidopsis thaliana         Arabidopsis\n",
       "3        Brassica oleracea           Boleracea\n",
       "4            Brassica rapa               Brapa\n",
       "5              Glycine max             Soybean\n",
       "6          Hordeum vulgare              Barley\n",
       "7      Medicago truncatula        Barrel medic\n",
       "8       Oropetium thomaeum  Resurrection grass\n",
       "9             Oryza sativa                Rice\n",
       "10    Oryza sativa Kitaake            Okitaake\n",
       "11          Panicum hallii   Hall's panicgrass\n",
       "12       Pharus latifolius              Pharus\n",
       "13      Phaseolus vulgaris         Common bean\n",
       "14    Solanum lycopersicum              Tomato\n",
       "15         Sorghum bicolor             Sorghum\n",
       "16  Thinopyrum intermedium          Wheatgrass\n",
       "17       Triticum aestivum               Wheat\n",
       "18       Vigna unguiculata              Cowpea\n",
       "19                Zea mays               Maize"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def show_available_species(con):\n",
    "    con = sqlite3.connect('SQNce.db')\n",
    "    cursorObj = con.cursor()\n",
    "    return(pd.read_sql_query(\"SELECT * FROM species\", con))\n",
    "\n",
    "show_available_species(con)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "devoted-matter",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "comparative-single",
   "metadata": {},
   "outputs": [],
   "source": [
    "cursorObj = con.cursor()\n",
    "cursorObj.execute('''SELECT gene_id, gene_annotation \n",
    "                     FROM gene_annotations \n",
    "                     WHERE gene_id =  ?  ''', (entity,))\n",
    "# (name,) - need the comma to treat it as a single item and not list of letters\n",
    "selected = cursorObj.fetchall()[0]\n",
    "od[selected[0]] = selected[1]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "missing-opera",
   "metadata": {},
   "outputs": [],
   "source": [
    "# TODO add documentation to SQNce queries\n",
    "\n",
    "def protein_seq_select(con, entity_list):\n",
    "    od = OrderedDict()\n",
    "    for entity in entity_list:\n",
    "        cursorObj = con.cursor()\n",
    "        cursorObj.execute('''SELECT protein_variant, gene_annotation \n",
    "                             FROM protein_seqs \n",
    "                             WHERE protein_variant =  ?  ''', (entity,))\n",
    "        # (name,) - need the comma to treat it as a single item and not list of letters\n",
    "        selected = cursorObj.fetchall()[0]\n",
    "        record = SeqRecord(Seq(zlib.decompress(selected[1]).decode(encoding='UTF-8')), \n",
    "                           id=selected[0], name=\"\", description=\"\")\n",
    "        od[selected[0]] = record\n",
    "        with open(\"selected.fasta\", 'w') as handle:\n",
    "            SeqIO.write(od.values(), handle, 'fasta')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "julian-bristol",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GeneID</th>\n",
       "      <th>annotation</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Zm00001d010294</td>\n",
       "      <td>Ubiquitin-associated/translation elongation fa...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>dsa</td>\n",
       "      <td>Gene not found</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Sobic.002G128101</td>\n",
       "      <td>protein kinase family protein</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AT2G34360</td>\n",
       "      <td>MATE efflux family protein, putative, expressed</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Sobic.002G128101</td>\n",
       "      <td>protein kinase family protein</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>002G128101</td>\n",
       "      <td>Gene not found</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             GeneID                                         annotation\n",
       "0    Zm00001d010294  Ubiquitin-associated/translation elongation fa...\n",
       "1               dsa                                     Gene not found\n",
       "2  Sobic.002G128101                      protein kinase family protein\n",
       "3         AT2G34360    MATE efflux family protein, putative, expressed\n",
       "4  Sobic.002G128101                      protein kinase family protein\n",
       "5        002G128101                                     Gene not found"
      ]
     },
     "execution_count": 36,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def annotation_select(con, entity_list):\n",
    "    ls = []\n",
    "    for entity in entity_list:\n",
    "        cursorObj = con.cursor()\n",
    "        cursorObj.execute('''SELECT gene_id, gene_annotation \n",
    "                             FROM gene_annotations \n",
    "                             WHERE gene_id =  ?  ''', (entity,))\n",
    "        # (name,) - need the comma to treat it as a single item and not list of letters\n",
    "        selected = cursorObj.fetchall()\n",
    "        if selected == []:\n",
    "            ls.append(\"Gene not found\")\n",
    "        else:\n",
    "            ls.append(selected[0][1])    \n",
    "    return(ls)\n",
    "\n",
    "con = sqlite3.connect('SQNce.db')\n",
    "gene_list = [\"Zm00001d010294\", \"dsa\", \"Sobic.002G128101\", \"AT2G34360\", \"Sobic.002G128101\", \"002G128101\"]\n",
    "df = pd.DataFrame({\"name\": gene_list, \"annotation\": annotation_select(con, gene_list) })\n",
    "df.columns = [\"GeneID\", \"annotation\"]\n",
    "df\n",
    "#pd.DataFrame.from_dict(annotation_select(con, [\"Zm00001d010294\", \"Sobic.002G128101\", \"AT2G34360\"], \"dict\"), \n",
    "#                       orient=\"index\", columns=[\"annotation\"])\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "endless-directive",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['Ubiquitin-associated/translation elongation factor EF1B protein', 'Gene not found', 'protein kinase family protein', 'MATE efflux family protein, putative, expressed']\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "['Ubiquitin-associated/translation elongation factor EF1B protein',\n",
       " 'Gene not found',\n",
       " 'protein kinase family protein',\n",
       " 'MATE efflux family protein, putative, expressed']"
      ]
     },
     "execution_count": 33,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "annotation_select(con, [\"Zm00001d010294\", \"Sobic.002G128101\", \"AT2G34360\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "id": "common-brown",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GeneID</th>\n",
       "      <th>annotation</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Zm00001d010294</td>\n",
       "      <td>Ubiquitin-associated/translation elongation fa...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Sobic.002G128101</td>\n",
       "      <td>protein kinase family protein</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AT2G34360</td>\n",
       "      <td>MATE efflux family protein, putative, expressed</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             GeneID                                         annotation\n",
       "0    Zm00001d010294  Ubiquitin-associated/translation elongation fa...\n",
       "1  Sobic.002G128101                      protein kinase family protein\n",
       "2         AT2G34360    MATE efflux family protein, putative, expressed"
      ]
     },
     "execution_count": 47,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "test = pd.DataFrame.from_dict(annotation_select(con, [\"Zm00001d010294\", \"Sobic.002G128101\", \"AT2G34360\"]), \n",
    "                              orient=\"index\").reset_index()\n",
    "test.columns = [\"GeneID\", \"annotation\"]\n",
    "test "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "satellite-newfoundland",
   "metadata": {},
   "source": [
    "# Example Queries"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "turkish-presence",
   "metadata": {},
   "outputs": [],
   "source": [
    "con = sqlite3.connect('SQNce.db')\n",
    "# Use the above query functions to parse SQNce with your gene lists \n",
    "input_value = [your_gene_list]\n",
    "protein_seq_select(con, input_value)\n",
    "con.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "seasonal-liquid",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[1, 2, 3, 3, 4, 5, 6]"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "[1,2,3] + [3,4,5,6]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "indie-supply",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[{'label': 1, 'value': 2}, {'label': 2, 'value': 3}]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "[{ 'label': label, 'value': val} for label, val in [[1,2], [2,3]]]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "everyday-freeware",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[{'label': 1, 'value': 2},\n",
       " {'label': 2, 'value': 3},\n",
       " {'label': 'Arabidopsis', 'value': 'Arabidopsis'},\n",
       " {'label': 'B73v2', 'value': 'B73v2'},\n",
       " {'label': 'B73v4', 'value': 'B73v4'},\n",
       " {'label': 'Gmax', 'value': 'Gmax'},\n",
       " {'label': 'Sorghum', 'value': 'Sorghum'}]"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def distinct_db_vals(db, table, column, custom_vals=[]):\n",
    "    # Input is the column to select and from which table\n",
    "    # Returns a list of all values in a specific table from SQNce.db\n",
    "    # Custom vals are added to the front using nested list of [label, value]\n",
    "    ls = [{ 'label': label, 'value': val} for label, val in custom_vals]\n",
    "    con = sqlite3.connect(db) # deploy with this\n",
    "    cursorObj = con.cursor()\n",
    "    distinct_df = pd.read_sql_query('''SELECT DISTINCT {0} \n",
    "                                       FROM {1}'''.format(column, table), con)\n",
    "    for name in distinct_df[column]:\n",
    "        ls.append({'label': name, 'value': name})\n",
    "    return(ls)\n",
    "distinct_db_vals(\"SQNce.db\", \"gene_coordinates\", \"genotype_id\",[[1,2], [2,3]])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "id": "spiritual-rugby",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['Cytochrome P450', 'Gene not found', 'Terpenoid synthases']"
      ]
     },
     "execution_count": 49,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def family_gene_select(gene_list):\n",
    "    # Use an input list of genes to find their family assignments\n",
    "    con = sqlite3.connect(\"SQNce.db\")\n",
    "    ls = []\n",
    "    for gene in gene_list:\n",
    "        cursorObj = con.cursor()\n",
    "        cursorObj.execute('''SELECT protein_id, family_name \n",
    "                            FROM gene_families\n",
    "                            WHERE protein_id =  ? ''', (gene,))\n",
    "        selected = cursorObj.fetchall()\n",
    "        if selected == []:\n",
    "            ls.append(\"Gene not found\")\n",
    "        else:\n",
    "            ls.append(selected[0][1])    \n",
    "    return(ls)\n",
    "family_gene_select([\"Seita.5G010100\", \"test\", \"Zm00001d011673\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "id": "correct-norfolk",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>protein_id</th>\n",
       "      <th>family_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Seita.5G085000</td>\n",
       "      <td>Cytochrome P450</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Seita.5G010200</td>\n",
       "      <td>Cytochrome P450</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Seita.5G010100</td>\n",
       "      <td>Cytochrome P450</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Seita.9G408200</td>\n",
       "      <td>Cytochrome P450</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Seita.9G011300</td>\n",
       "      <td>Cytochrome P450</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>833</th>\n",
       "      <td>Zm00001d011673</td>\n",
       "      <td>Terpenoid synthases</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>834</th>\n",
       "      <td>Zm00001d009431</td>\n",
       "      <td>Terpenoid synthases</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>835</th>\n",
       "      <td>Zm00001d050159</td>\n",
       "      <td>Terpenoid synthases</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>836</th>\n",
       "      <td>Zm00001d014367</td>\n",
       "      <td>Terpenoid synthases</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>837</th>\n",
       "      <td>Zm00001d014369</td>\n",
       "      <td>Terpenoid synthases</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>838 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         protein_id          family_name\n",
       "0    Seita.5G085000      Cytochrome P450\n",
       "1    Seita.5G010200      Cytochrome P450\n",
       "2    Seita.5G010100      Cytochrome P450\n",
       "3    Seita.9G408200      Cytochrome P450\n",
       "4    Seita.9G011300      Cytochrome P450\n",
       "..              ...                  ...\n",
       "833  Zm00001d011673  Terpenoid synthases\n",
       "834  Zm00001d009431  Terpenoid synthases\n",
       "835  Zm00001d050159  Terpenoid synthases\n",
       "836  Zm00001d014367  Terpenoid synthases\n",
       "837  Zm00001d014369  Terpenoid synthases\n",
       "\n",
       "[838 rows x 2 columns]"
      ]
     },
     "execution_count": 45,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "con = sqlite3.connect(\"SQNce.db\") # deploy with this\n",
    "cursorObj = con.cursor()\n",
    "genotype = str(\"','\".join(['Zea mays', \"dsa\",'Setaria italica']))\n",
    "family = str(\"','\".join(['Terpenoid synthases', 'Cytochrome P450']))\n",
    "df = pd.read_sql_query(\"\"\"SELECT protein_id, family_name \n",
    "                                   FROM gene_families\n",
    "                                   WHERE species_id IN ('{0}') AND family_name IN ('{1}')\"\"\".format(genotype, family), con)\n",
    "df"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
