{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "approved-horizon",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import gzip\n",
    "import sqlite3\n",
    "import zlib\n",
    "from sqlite3 import Error\n",
    "import pandas as pd\n",
    "from collections import OrderedDict\n",
    "\n",
    "from Bio import SeqIO \n",
    "from Bio.SeqRecord import SeqRecord\n",
    "from Bio.Seq import Seq"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "quarterly-slovenia",
   "metadata": {},
   "source": [
    "# SQNce Creation Functions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "contrary-writer",
   "metadata": {},
   "outputs": [],
   "source": [
    "# TODO add documentation to all SQNce creation functions\n",
    "\n",
    "# Establish connection with SQNce.db, generating a new SQLite3 database if needed\n",
    "def sql_connection():\n",
    "    try:\n",
    "        con = sqlite3.connect('SQNce-proteomes.db')\n",
    "        print(\"Connection established.\")\n",
    "        return(con)\n",
    "    except Error:\n",
    "        print(Error)\n",
    "\n",
    "# After establishing connection with SQNce create the specified tables\n",
    "def sql_table(con):\n",
    "    cursorObj = con.cursor()\n",
    "    cursorObj.execute(\"\"\"CREATE TABLE IF NOT EXISTS protein_seqs(\n",
    "                         protein_id text,\n",
    "                         species_id text,\n",
    "                         genotype_id text,\n",
    "                         protein_length text,\n",
    "                         protein_sequence blob,\n",
    "                         FOREIGN KEY (genotype_id) REFERENCES genotypes(genotype_id)) \n",
    "                         \"\"\")   \n",
    "    con.commit()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "available-seeking",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Connection established.\n"
     ]
    }
   ],
   "source": [
    "# Current implementation requires re-parsing of all the input files to create SQNce\n",
    "# TODO SQNce update functions to parse input data only if not previously included \n",
    "if os.path.exists(\"SQNce-proteomes.db\"): os.remove(\"SQNce-proteomes.db\")\n",
    "con = sql_connection()\n",
    "sql_table(con)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "eleven-utility",
   "metadata": {},
   "source": [
    "# SQNce Data Input Functions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "noted-dallas",
   "metadata": {},
   "outputs": [],
   "source": [
    "# TODO add documentation to all SQNce Data Input Functions\n",
    "    \n",
    "# https://stackoverflow.com/questions/18219779/bulk-insert-huge-data-into-sqlite-using-python\n",
    "def protein_seq_insert(con, entity_list):\n",
    "    cursorObj = con.cursor()\n",
    "    cursorObj.executemany(\"\"\"INSERT INTO protein_seqs(\n",
    "                         protein_id,\n",
    "                         species_id,\n",
    "                         genotype_id, \n",
    "                         protein_length,  \n",
    "                         protein_sequence) \n",
    "                         VALUES(?, ?, ?, ?, ?)\"\"\", entity_list)\n",
    "    con.commit()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "developing-tribe",
   "metadata": {},
   "outputs": [],
   "source": [
    "#######################################################################\n",
    "########################### Protein Seqs ##############################\n",
    "#######################################################################\n",
    "df = pd.read_csv(\"init/proteins-db.tsv\", sep=\"\\t\")\n",
    "for index, row in df.iterrows():\n",
    "    fasta_file = gzip.open(\"inputs/proteins/\"+row[2], mode='rt')\n",
    "    protein_seq_list = []\n",
    "    for seq in SeqIO.parse(fasta_file, \"fasta\"):\n",
    "        # Protein sequences are saved as as a Binary data type for compression\n",
    "        protein_seq_list.append([seq.id, row[0], row[1], len(seq.seq), \n",
    "                                 sqlite3.Binary(zlib.compress(str(seq.seq).encode('utf-8')))])\n",
    "    protein_seq_insert(con, protein_seq_list)\n",
    "# Create keys for protein IDs with isoform and without\n",
    "cursorObj = con.cursor()\n",
    "cursorObj.execute(\"CREATE INDEX proteins_ids_index ON protein_seqs(protein_id)\")\n",
    "con.commit()    "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "timely-still",
   "metadata": {},
   "source": [
    "# SQNce Query Functions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "intense-threat",
   "metadata": {},
   "outputs": [
    {
     "ename": "SyntaxError",
     "evalue": "'return' outside function (<ipython-input-27-5da10337c74d>, line 7)",
     "output_type": "error",
     "traceback": [
      "\u001b[1;36m  File \u001b[1;32m\"<ipython-input-27-5da10337c74d>\"\u001b[1;36m, line \u001b[1;32m7\u001b[0m\n\u001b[1;33m    return(html.P(\"Number of columns is not 2. Use tab-seperated values.\"))\u001b[0m\n\u001b[1;37m    ^\u001b[0m\n\u001b[1;31mSyntaxError\u001b[0m\u001b[1;31m:\u001b[0m 'return' outside function\n"
     ]
    }
   ],
   "source": [
    "rows = \"\"\"Chr2\\t12345\\nChr3\\t54354\\nChr2\\t5234354\\n\"\"\".split(\"\\n\")\n",
    "coordinate_list = [row.split(\"\\t\") for row in rows]\n",
    "if coordinate_list[-1]==[\"\"]:\n",
    "    coordinate_list = coordinate_list[:-1]\n",
    "for row in coordinate_list:\n",
    "    if len(row) != 2:\n",
    "        return(html.P(\"Number of columns is not 2. Use tab-seperated values.\"))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "id": "false-craps",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Query to find neighboring genes\n",
    "def get_SNP_neighbors(genotype, chromsome, coordinate, distance):\n",
    "    con = sqlite3.connect('SQNce.db')\n",
    "    cursorObj = con.cursor()\n",
    "    df = pd.read_sql_query('''SELECT * \n",
    "                     FROM gene_coordinates \n",
    "                     WHERE genotype_id = \"{0}\"\n",
    "                     AND gene_chr = \"{1}\"\n",
    "                     AND gene_start BETWEEN {2} AND {3}\n",
    "                     \n",
    "                     UNION ALL\n",
    "                     \n",
    "                     SELECT * \n",
    "                     FROM gene_coordinates \n",
    "                     WHERE genotype_id = \"{0}\"\n",
    "                     AND gene_chr = \"{1}\"\n",
    "                     AND gene_start BETWEEN {2} AND {3}\n",
    "                     '''.format(genotype, chromsome, coordinate-distance, coordinate+distance), con)\n",
    "    # Should check why it returns the same row twice, probably need to correct the query\n",
    "    df = df.drop_duplicates()\n",
    "    df.insert(0, 'Query', pd.Series([\"_\".join([chromsome, str(coordinate)]) for x in range(len(df.index))]))\n",
    "    return(df)\n",
    "\n",
    "df = pd.DataFrame()\n",
    "for entity in [[\"Chr2\", 19681637], [\"Chr2\", 1234564], [\"Chr4\", 1234564], [\"Chr2\", 19681638]]:\n",
    "      df = pd.concat([df, get_SNP_neighbors(\"Arabidopsis\", entity[0], entity[1], 10000)])  \n",
    "#df.drop_duplicates(subset=[\"gene_id\"]).reset_index()\n",
    "#df1 = get_SNP_neighbors(\"Arabidopsis\", \"Chr2\", 19681637, 10000)\n",
    "#df1\n",
    "#pd.concat([df1, df1])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "id": "centered-script",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Query</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>genotype_id</th>\n",
       "      <th>gene_chr</th>\n",
       "      <th>gene_start</th>\n",
       "      <th>gene_end</th>\n",
       "      <th>gene_orientation</th>\n",
       "      <th>annotation</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Chr2_19681637</td>\n",
       "      <td>AT2G48110</td>\n",
       "      <td>Arabidopsis</td>\n",
       "      <td>Chr2</td>\n",
       "      <td>19673293</td>\n",
       "      <td>19679711</td>\n",
       "      <td>+</td>\n",
       "      <td>structural constituent of ribosome, putative, ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Chr2_19681637</td>\n",
       "      <td>AT2G48120</td>\n",
       "      <td>Arabidopsis</td>\n",
       "      <td>Chr2</td>\n",
       "      <td>19679730</td>\n",
       "      <td>19681546</td>\n",
       "      <td>+</td>\n",
       "      <td>PAC, putative, expressed</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Chr2_19681637</td>\n",
       "      <td>AT2G48121</td>\n",
       "      <td>Arabidopsis</td>\n",
       "      <td>Chr2</td>\n",
       "      <td>19681637</td>\n",
       "      <td>19682391</td>\n",
       "      <td>+</td>\n",
       "      <td>chloroplast ribonuclease III domain protein, p...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Chr2_19681637</td>\n",
       "      <td>AT2G48130</td>\n",
       "      <td>Arabidopsis</td>\n",
       "      <td>Chr2</td>\n",
       "      <td>19685066</td>\n",
       "      <td>19685993</td>\n",
       "      <td>-</td>\n",
       "      <td>LTPL78 - Protease inhibitor/seed storage/LTP f...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Chr2_19681637</td>\n",
       "      <td>AT2G48140</td>\n",
       "      <td>Arabidopsis</td>\n",
       "      <td>Chr2</td>\n",
       "      <td>19686409</td>\n",
       "      <td>19687664</td>\n",
       "      <td>+</td>\n",
       "      <td>LTPL99 - Protease inhibitor/seed storage/LTP f...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>Chr2_19681637</td>\n",
       "      <td>AT2G48150</td>\n",
       "      <td>Arabidopsis</td>\n",
       "      <td>Chr2</td>\n",
       "      <td>19687962</td>\n",
       "      <td>19689173</td>\n",
       "      <td>-</td>\n",
       "      <td>glutathione peroxidase domain containing prote...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>Chr2_19681637</td>\n",
       "      <td>AT2G48160</td>\n",
       "      <td>Arabidopsis</td>\n",
       "      <td>Chr2</td>\n",
       "      <td>19689409</td>\n",
       "      <td>19696821</td>\n",
       "      <td>-</td>\n",
       "      <td>PWWP domain containing protein, expressed</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Chr2_1234564</td>\n",
       "      <td>AT2G03955</td>\n",
       "      <td>Arabidopsis</td>\n",
       "      <td>Chr2</td>\n",
       "      <td>1238677</td>\n",
       "      <td>1239071</td>\n",
       "      <td>+</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Chr4_1234564</td>\n",
       "      <td>AT4G02770</td>\n",
       "      <td>Arabidopsis</td>\n",
       "      <td>Chr4</td>\n",
       "      <td>1229111</td>\n",
       "      <td>1229945</td>\n",
       "      <td>-</td>\n",
       "      <td>photosystem I reaction center subunit II, chlo...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Chr4_1234564</td>\n",
       "      <td>AT4G02780</td>\n",
       "      <td>Arabidopsis</td>\n",
       "      <td>Chr4</td>\n",
       "      <td>1237767</td>\n",
       "      <td>1244813</td>\n",
       "      <td>-</td>\n",
       "      <td>ent-kaurene synthase, chloroplast precursor, p...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Chr2_19681638</td>\n",
       "      <td>AT2G48110</td>\n",
       "      <td>Arabidopsis</td>\n",
       "      <td>Chr2</td>\n",
       "      <td>19673293</td>\n",
       "      <td>19679711</td>\n",
       "      <td>+</td>\n",
       "      <td>structural constituent of ribosome, putative, ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Chr2_19681638</td>\n",
       "      <td>AT2G48120</td>\n",
       "      <td>Arabidopsis</td>\n",
       "      <td>Chr2</td>\n",
       "      <td>19679730</td>\n",
       "      <td>19681546</td>\n",
       "      <td>+</td>\n",
       "      <td>PAC, putative, expressed</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Chr2_19681638</td>\n",
       "      <td>AT2G48121</td>\n",
       "      <td>Arabidopsis</td>\n",
       "      <td>Chr2</td>\n",
       "      <td>19681637</td>\n",
       "      <td>19682391</td>\n",
       "      <td>+</td>\n",
       "      <td>chloroplast ribonuclease III domain protein, p...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Chr2_19681638</td>\n",
       "      <td>AT2G48130</td>\n",
       "      <td>Arabidopsis</td>\n",
       "      <td>Chr2</td>\n",
       "      <td>19685066</td>\n",
       "      <td>19685993</td>\n",
       "      <td>-</td>\n",
       "      <td>LTPL78 - Protease inhibitor/seed storage/LTP f...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Chr2_19681638</td>\n",
       "      <td>AT2G48140</td>\n",
       "      <td>Arabidopsis</td>\n",
       "      <td>Chr2</td>\n",
       "      <td>19686409</td>\n",
       "      <td>19687664</td>\n",
       "      <td>+</td>\n",
       "      <td>LTPL99 - Protease inhibitor/seed storage/LTP f...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>Chr2_19681638</td>\n",
       "      <td>AT2G48150</td>\n",
       "      <td>Arabidopsis</td>\n",
       "      <td>Chr2</td>\n",
       "      <td>19687962</td>\n",
       "      <td>19689173</td>\n",
       "      <td>-</td>\n",
       "      <td>glutathione peroxidase domain containing prote...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>Chr2_19681638</td>\n",
       "      <td>AT2G48160</td>\n",
       "      <td>Arabidopsis</td>\n",
       "      <td>Chr2</td>\n",
       "      <td>19689409</td>\n",
       "      <td>19696821</td>\n",
       "      <td>-</td>\n",
       "      <td>PWWP domain containing protein, expressed</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           Query    gene_id  genotype_id gene_chr  gene_start  gene_end  \\\n",
       "0  Chr2_19681637  AT2G48110  Arabidopsis     Chr2    19673293  19679711   \n",
       "1  Chr2_19681637  AT2G48120  Arabidopsis     Chr2    19679730  19681546   \n",
       "2  Chr2_19681637  AT2G48121  Arabidopsis     Chr2    19681637  19682391   \n",
       "3  Chr2_19681637  AT2G48130  Arabidopsis     Chr2    19685066  19685993   \n",
       "4  Chr2_19681637  AT2G48140  Arabidopsis     Chr2    19686409  19687664   \n",
       "5  Chr2_19681637  AT2G48150  Arabidopsis     Chr2    19687962  19689173   \n",
       "6  Chr2_19681637  AT2G48160  Arabidopsis     Chr2    19689409  19696821   \n",
       "0   Chr2_1234564  AT2G03955  Arabidopsis     Chr2     1238677   1239071   \n",
       "0   Chr4_1234564  AT4G02770  Arabidopsis     Chr4     1229111   1229945   \n",
       "1   Chr4_1234564  AT4G02780  Arabidopsis     Chr4     1237767   1244813   \n",
       "0  Chr2_19681638  AT2G48110  Arabidopsis     Chr2    19673293  19679711   \n",
       "1  Chr2_19681638  AT2G48120  Arabidopsis     Chr2    19679730  19681546   \n",
       "2  Chr2_19681638  AT2G48121  Arabidopsis     Chr2    19681637  19682391   \n",
       "3  Chr2_19681638  AT2G48130  Arabidopsis     Chr2    19685066  19685993   \n",
       "4  Chr2_19681638  AT2G48140  Arabidopsis     Chr2    19686409  19687664   \n",
       "5  Chr2_19681638  AT2G48150  Arabidopsis     Chr2    19687962  19689173   \n",
       "6  Chr2_19681638  AT2G48160  Arabidopsis     Chr2    19689409  19696821   \n",
       "\n",
       "  gene_orientation                                         annotation  \n",
       "0                +  structural constituent of ribosome, putative, ...  \n",
       "1                +                           PAC, putative, expressed  \n",
       "2                +  chloroplast ribonuclease III domain protein, p...  \n",
       "3                -  LTPL78 - Protease inhibitor/seed storage/LTP f...  \n",
       "4                +  LTPL99 - Protease inhibitor/seed storage/LTP f...  \n",
       "5                -  glutathione peroxidase domain containing prote...  \n",
       "6                -          PWWP domain containing protein, expressed  \n",
       "0                +                                               None  \n",
       "0                -  photosystem I reaction center subunit II, chlo...  \n",
       "1                -  ent-kaurene synthase, chloroplast precursor, p...  \n",
       "0                +  structural constituent of ribosome, putative, ...  \n",
       "1                +                           PAC, putative, expressed  \n",
       "2                +  chloroplast ribonuclease III domain protein, p...  \n",
       "3                -  LTPL78 - Protease inhibitor/seed storage/LTP f...  \n",
       "4                +  LTPL99 - Protease inhibitor/seed storage/LTP f...  \n",
       "5                -  glutathione peroxidase domain containing prote...  \n",
       "6                -          PWWP domain containing protein, expressed  "
      ]
     },
     "execution_count": 39,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Query to find neighboring genes and annotate them\n",
    "\n",
    "df[\"annotation\"] = annotation_select(con, df[\"gene_id\"].to_list())\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "according-closer",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>species_name</th>\n",
       "      <th>common_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Arabidopsis halleri</td>\n",
       "      <td>Ahalleri</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Arabidopsis lyrata</td>\n",
       "      <td>Alyrata</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Arabidopsis thaliana</td>\n",
       "      <td>Arabidopsis</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Brassica oleracea</td>\n",
       "      <td>Boleracea</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Brassica rapa</td>\n",
       "      <td>Brapa</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>Glycine max</td>\n",
       "      <td>Soybean</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>Hordeum vulgare</td>\n",
       "      <td>Barley</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>Medicago truncatula</td>\n",
       "      <td>Barrel medic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>Oropetium thomaeum</td>\n",
       "      <td>Resurrection grass</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>Oryza sativa</td>\n",
       "      <td>Rice</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>Oryza sativa Kitaake</td>\n",
       "      <td>Okitaake</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>Panicum hallii</td>\n",
       "      <td>Hall's panicgrass</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>Pharus latifolius</td>\n",
       "      <td>Pharus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>Phaseolus vulgaris</td>\n",
       "      <td>Common bean</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>Solanum lycopersicum</td>\n",
       "      <td>Tomato</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>Sorghum bicolor</td>\n",
       "      <td>Sorghum</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>Thinopyrum intermedium</td>\n",
       "      <td>Wheatgrass</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>Triticum aestivum</td>\n",
       "      <td>Wheat</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>Vigna unguiculata</td>\n",
       "      <td>Cowpea</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>Zea mays</td>\n",
       "      <td>Maize</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "              species_name         common_name\n",
       "0      Arabidopsis halleri            Ahalleri\n",
       "1       Arabidopsis lyrata             Alyrata\n",
       "2     Arabidopsis thaliana         Arabidopsis\n",
       "3        Brassica oleracea           Boleracea\n",
       "4            Brassica rapa               Brapa\n",
       "5              Glycine max             Soybean\n",
       "6          Hordeum vulgare              Barley\n",
       "7      Medicago truncatula        Barrel medic\n",
       "8       Oropetium thomaeum  Resurrection grass\n",
       "9             Oryza sativa                Rice\n",
       "10    Oryza sativa Kitaake            Okitaake\n",
       "11          Panicum hallii   Hall's panicgrass\n",
       "12       Pharus latifolius              Pharus\n",
       "13      Phaseolus vulgaris         Common bean\n",
       "14    Solanum lycopersicum              Tomato\n",
       "15         Sorghum bicolor             Sorghum\n",
       "16  Thinopyrum intermedium          Wheatgrass\n",
       "17       Triticum aestivum               Wheat\n",
       "18       Vigna unguiculata              Cowpea\n",
       "19                Zea mays               Maize"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def show_available_species(con):\n",
    "    con = sqlite3.connect('SQNce.db')\n",
    "    cursorObj = con.cursor()\n",
    "    return(pd.read_sql_query(\"SELECT * FROM species\", con))\n",
    "\n",
    "show_available_species(con)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "buried-saudi",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bright-diagnosis",
   "metadata": {},
   "outputs": [],
   "source": [
    "cursorObj = con.cursor()\n",
    "cursorObj.execute('''SELECT gene_id, gene_annotation \n",
    "                     FROM gene_annotations \n",
    "                     WHERE gene_id =  ?  ''', (entity,))\n",
    "# (name,) - need the comma to treat it as a single item and not list of letters\n",
    "selected = cursorObj.fetchall()[0]\n",
    "od[selected[0]] = selected[1]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "labeled-pollution",
   "metadata": {},
   "outputs": [],
   "source": [
    "# TODO add documentation to SQNce queries\n",
    "\n",
    "def protein_seq_select(con, entity_list):\n",
    "    od = OrderedDict()\n",
    "    for entity in entity_list:\n",
    "        cursorObj = con.cursor()\n",
    "        cursorObj.execute('''SELECT protein_variant, gene_annotation \n",
    "                             FROM protein_seqs \n",
    "                             WHERE protein_variant =  ?  ''', (entity,))\n",
    "        # (name,) - need the comma to treat it as a single item and not list of letters\n",
    "        selected = cursorObj.fetchall()[0]\n",
    "        record = SeqRecord(Seq(zlib.decompress(selected[1]).decode(encoding='UTF-8')), \n",
    "                           id=selected[0], name=\"\", description=\"\")\n",
    "        od[selected[0]] = record\n",
    "        with open(\"selected.fasta\", 'w') as handle:\n",
    "            SeqIO.write(od.values(), handle, 'fasta')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "collective-tribune",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GeneID</th>\n",
       "      <th>annotation</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Zm00001d010294</td>\n",
       "      <td>Ubiquitin-associated/translation elongation fa...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>dsa</td>\n",
       "      <td>Gene not found</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Sobic.002G128101</td>\n",
       "      <td>protein kinase family protein</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AT2G34360</td>\n",
       "      <td>MATE efflux family protein, putative, expressed</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Sobic.002G128101</td>\n",
       "      <td>protein kinase family protein</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>002G128101</td>\n",
       "      <td>Gene not found</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             GeneID                                         annotation\n",
       "0    Zm00001d010294  Ubiquitin-associated/translation elongation fa...\n",
       "1               dsa                                     Gene not found\n",
       "2  Sobic.002G128101                      protein kinase family protein\n",
       "3         AT2G34360    MATE efflux family protein, putative, expressed\n",
       "4  Sobic.002G128101                      protein kinase family protein\n",
       "5        002G128101                                     Gene not found"
      ]
     },
     "execution_count": 36,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def annotation_select(con, entity_list):\n",
    "    ls = []\n",
    "    for entity in entity_list:\n",
    "        cursorObj = con.cursor()\n",
    "        cursorObj.execute('''SELECT gene_id, gene_annotation \n",
    "                             FROM gene_annotations \n",
    "                             WHERE gene_id =  ?  ''', (entity,))\n",
    "        # (name,) - need the comma to treat it as a single item and not list of letters\n",
    "        selected = cursorObj.fetchall()\n",
    "        if selected == []:\n",
    "            ls.append(\"Gene not found\")\n",
    "        else:\n",
    "            ls.append(selected[0][1])    \n",
    "    return(ls)\n",
    "\n",
    "con = sqlite3.connect('SQNce.db')\n",
    "gene_list = [\"Zm00001d010294\", \"dsa\", \"Sobic.002G128101\", \"AT2G34360\", \"Sobic.002G128101\", \"002G128101\"]\n",
    "df = pd.DataFrame({\"name\": gene_list, \"annotation\": annotation_select(con, gene_list) })\n",
    "df.columns = [\"GeneID\", \"annotation\"]\n",
    "df\n",
    "#pd.DataFrame.from_dict(annotation_select(con, [\"Zm00001d010294\", \"Sobic.002G128101\", \"AT2G34360\"], \"dict\"), \n",
    "#                       orient=\"index\", columns=[\"annotation\"])\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "smart-variance",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['Ubiquitin-associated/translation elongation factor EF1B protein', 'Gene not found', 'protein kinase family protein', 'MATE efflux family protein, putative, expressed']\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "['Ubiquitin-associated/translation elongation factor EF1B protein',\n",
       " 'Gene not found',\n",
       " 'protein kinase family protein',\n",
       " 'MATE efflux family protein, putative, expressed']"
      ]
     },
     "execution_count": 33,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "annotation_select(con, [\"Zm00001d010294\", \"Sobic.002G128101\", \"AT2G34360\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "id": "presidential-drinking",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GeneID</th>\n",
       "      <th>annotation</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Zm00001d010294</td>\n",
       "      <td>Ubiquitin-associated/translation elongation fa...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Sobic.002G128101</td>\n",
       "      <td>protein kinase family protein</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AT2G34360</td>\n",
       "      <td>MATE efflux family protein, putative, expressed</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             GeneID                                         annotation\n",
       "0    Zm00001d010294  Ubiquitin-associated/translation elongation fa...\n",
       "1  Sobic.002G128101                      protein kinase family protein\n",
       "2         AT2G34360    MATE efflux family protein, putative, expressed"
      ]
     },
     "execution_count": 47,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "test = pd.DataFrame.from_dict(annotation_select(con, [\"Zm00001d010294\", \"Sobic.002G128101\", \"AT2G34360\"]), \n",
    "                              orient=\"index\").reset_index()\n",
    "test.columns = [\"GeneID\", \"annotation\"]\n",
    "test "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fabulous-consumption",
   "metadata": {},
   "source": [
    "# Example Queries"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "postal-possession",
   "metadata": {},
   "outputs": [],
   "source": [
    "con = sqlite3.connect('SQNce.db')\n",
    "# Use the above query functions to parse SQNce with your gene lists \n",
    "input_value = [your_gene_list]\n",
    "protein_seq_select(con, input_value)\n",
    "con.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "processed-census",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[{'label': 1, 'value': 2}, {'label': 2, 'value': 3}]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "[{ 'label': label, 'value': val} for label, val in [[1,2], [2,3]]]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "id": "scientific-heating",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['Arabidopsis', 'B73v2', 'B73v4', 'Gmax', 'Sorghum']"
      ]
     },
     "execution_count": 55,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def distinct_db_vals(db, table, column, custom_vals=[], return_ls=False):\n",
    "    # Input is the column to select and from which table\n",
    "    # Returns a list of all values in a specific table from SQNce.db\n",
    "    # Custom vals are added to the front using nested list of [label, value]\n",
    "    ls = [{ 'label': label, 'value': val} for label, val in custom_vals]\n",
    "    con = sqlite3.connect(db) # deploy with this\n",
    "    cursorObj = con.cursor()\n",
    "    distinct_df = pd.read_sql_query('''SELECT DISTINCT {0} \n",
    "                                       FROM {1}'''.format(column, table), con)\n",
    "    if return_ls:\n",
    "            return(distinct_df[column].to_list())\n",
    "    for name in distinct_df[column]:\n",
    "        ls.append({'label': name, 'value': name})\n",
    "    return(ls)\n",
    "distinct_db_vals(\"SQNce.db\", \"gene_coordinates\", \"genotype_id\",[[1,2], [2,3]], True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "id": "specialized-accent",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['Cytochrome P450', 'Gene not found', 'Terpenoid synthases']"
      ]
     },
     "execution_count": 49,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def family_gene_select(gene_list):\n",
    "    # Use an input list of genes to find their family assignments\n",
    "    con = sqlite3.connect(\"SQNce.db\")\n",
    "    ls = []\n",
    "    for gene in gene_list:\n",
    "        cursorObj = con.cursor()\n",
    "        cursorObj.execute('''SELECT protein_id, family_name \n",
    "                            FROM gene_families\n",
    "                            WHERE protein_id =  ? ''', (gene,))\n",
    "        selected = cursorObj.fetchall()\n",
    "        if selected == []:\n",
    "            ls.append(\"Gene not found\")\n",
    "        else:\n",
    "            ls.append(selected[0][1])    \n",
    "    return(ls)\n",
    "family_gene_select([\"Seita.5G010100\", \"test\", \"Zm00001d011673\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "id": "downtown-contract",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>protein_id</th>\n",
       "      <th>family_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Seita.5G085000</td>\n",
       "      <td>Cytochrome P450</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Seita.5G010200</td>\n",
       "      <td>Cytochrome P450</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Seita.5G010100</td>\n",
       "      <td>Cytochrome P450</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Seita.9G408200</td>\n",
       "      <td>Cytochrome P450</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Seita.9G011300</td>\n",
       "      <td>Cytochrome P450</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>833</th>\n",
       "      <td>Zm00001d011673</td>\n",
       "      <td>Terpenoid synthases</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>834</th>\n",
       "      <td>Zm00001d009431</td>\n",
       "      <td>Terpenoid synthases</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>835</th>\n",
       "      <td>Zm00001d050159</td>\n",
       "      <td>Terpenoid synthases</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>836</th>\n",
       "      <td>Zm00001d014367</td>\n",
       "      <td>Terpenoid synthases</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>837</th>\n",
       "      <td>Zm00001d014369</td>\n",
       "      <td>Terpenoid synthases</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>838 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         protein_id          family_name\n",
       "0    Seita.5G085000      Cytochrome P450\n",
       "1    Seita.5G010200      Cytochrome P450\n",
       "2    Seita.5G010100      Cytochrome P450\n",
       "3    Seita.9G408200      Cytochrome P450\n",
       "4    Seita.9G011300      Cytochrome P450\n",
       "..              ...                  ...\n",
       "833  Zm00001d011673  Terpenoid synthases\n",
       "834  Zm00001d009431  Terpenoid synthases\n",
       "835  Zm00001d050159  Terpenoid synthases\n",
       "836  Zm00001d014367  Terpenoid synthases\n",
       "837  Zm00001d014369  Terpenoid synthases\n",
       "\n",
       "[838 rows x 2 columns]"
      ]
     },
     "execution_count": 52,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "con = sqlite3.connect(\"SQNce.db\") # deploy with this\n",
    "cursorObj = con.cursor()\n",
    "genotype = str(\"','\".join(['Zea mays', \"dsa\",'Setaria italica']))\n",
    "family = str(\"','\".join(['Terpenoid synthases', 'Cytochrome P450']))\n",
    "df = pd.read_sql_query(\"\"\"SELECT protein_id, family_name \n",
    "                                   FROM gene_families\n",
    "                                   WHERE species_id IN ('{0}') AND family_name IN ('{1}')\"\"\".format(genotype, family), con)\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "conditional-journalist",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>subject_id</th>\n",
       "      <th>query_id</th>\n",
       "      <th>bit_score</th>\n",
       "      <th>subject_genotype</th>\n",
       "      <th>query_genotype</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Zm00001d014121</td>\n",
       "      <td>Zm00023ab228620</td>\n",
       "      <td>1005</td>\n",
       "      <td>B73v4</td>\n",
       "      <td>CML247</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Zm00001d014121</td>\n",
       "      <td>Zm00023ab144550</td>\n",
       "      <td>390</td>\n",
       "      <td>B73v4</td>\n",
       "      <td>CML247</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Zm00001d014121</td>\n",
       "      <td>Zm00023ab142020</td>\n",
       "      <td>357</td>\n",
       "      <td>B73v4</td>\n",
       "      <td>CML247</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Zm00001d014121</td>\n",
       "      <td>Zm00023ab171210</td>\n",
       "      <td>342</td>\n",
       "      <td>B73v4</td>\n",
       "      <td>CML247</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Zm00001d014121</td>\n",
       "      <td>Zm00023ab228750</td>\n",
       "      <td>733</td>\n",
       "      <td>B73v4</td>\n",
       "      <td>CML247</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1043</th>\n",
       "      <td>Zm00001d014121</td>\n",
       "      <td>Glyma.14G117200</td>\n",
       "      <td>409</td>\n",
       "      <td>B73v4</td>\n",
       "      <td>Wm82</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1044</th>\n",
       "      <td>Zm00001d014121</td>\n",
       "      <td>Glyma.14G015100</td>\n",
       "      <td>384</td>\n",
       "      <td>B73v4</td>\n",
       "      <td>Wm82</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1045</th>\n",
       "      <td>Zm00001d014121</td>\n",
       "      <td>Glyma.18G080400</td>\n",
       "      <td>370</td>\n",
       "      <td>B73v4</td>\n",
       "      <td>Wm82</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1046</th>\n",
       "      <td>Zm00001d014121</td>\n",
       "      <td>Glyma.18G080200</td>\n",
       "      <td>384</td>\n",
       "      <td>B73v4</td>\n",
       "      <td>Wm82</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1047</th>\n",
       "      <td>Zm00001d014121</td>\n",
       "      <td>Glyma.18G080300</td>\n",
       "      <td>393</td>\n",
       "      <td>B73v4</td>\n",
       "      <td>Wm82</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1048 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          subject_id         query_id  bit_score subject_genotype  \\\n",
       "0     Zm00001d014121  Zm00023ab228620       1005            B73v4   \n",
       "1     Zm00001d014121  Zm00023ab144550        390            B73v4   \n",
       "2     Zm00001d014121  Zm00023ab142020        357            B73v4   \n",
       "3     Zm00001d014121  Zm00023ab171210        342            B73v4   \n",
       "4     Zm00001d014121  Zm00023ab228750        733            B73v4   \n",
       "...              ...              ...        ...              ...   \n",
       "1043  Zm00001d014121  Glyma.14G117200        409            B73v4   \n",
       "1044  Zm00001d014121  Glyma.14G015100        384            B73v4   \n",
       "1045  Zm00001d014121  Glyma.18G080400        370            B73v4   \n",
       "1046  Zm00001d014121  Glyma.18G080200        384            B73v4   \n",
       "1047  Zm00001d014121  Glyma.18G080300        393            B73v4   \n",
       "\n",
       "     query_genotype  \n",
       "0            CML247  \n",
       "1            CML247  \n",
       "2            CML247  \n",
       "3            CML247  \n",
       "4            CML247  \n",
       "...             ...  \n",
       "1043           Wm82  \n",
       "1044           Wm82  \n",
       "1045           Wm82  \n",
       "1046           Wm82  \n",
       "1047           Wm82  \n",
       "\n",
       "[1048 rows x 5 columns]"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "selected = \"subject_id\"\n",
    "entity_list = [\"Zm00001d014121\", \"Zm00001d014134, \"\"Zm00001d014136\"]\n",
    "con = sqlite3.connect(\"SQNce.db\") # deploy with this\n",
    "cursorObj = con.cursor()\n",
    "entity_list_str = str(\"','\".join(entity_list))\n",
    "df = pd.read_sql_query(\"\"\"SELECT * \n",
    "                        FROM BBHs\n",
    "                        WHERE {0} IN ('{1}')\"\"\".format(selected, entity_list_str), con)\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "promising-prompt",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>subject_id</th>\n",
       "      <th>query_genotype</th>\n",
       "      <th>bit_score</th>\n",
       "      <th>query_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Zm00001d014121</td>\n",
       "      <td>B73v4</td>\n",
       "      <td>1015</td>\n",
       "      <td>Zm00001d014121</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Zm00001d014121</td>\n",
       "      <td>B73v5</td>\n",
       "      <td>1013</td>\n",
       "      <td>Zm00001eb222540</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Zm00001d014121</td>\n",
       "      <td>B97</td>\n",
       "      <td>1007</td>\n",
       "      <td>Zm00018ab231330</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Zm00001d014121</td>\n",
       "      <td>CML103</td>\n",
       "      <td>1012</td>\n",
       "      <td>Zm00021ab225420</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Zm00001d014121</td>\n",
       "      <td>CML228</td>\n",
       "      <td>1012</td>\n",
       "      <td>Zm00022ab224970</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>Zm00001d014121</td>\n",
       "      <td>CML247</td>\n",
       "      <td>1005</td>\n",
       "      <td>Zm00023ab228620</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>Zm00001d014121</td>\n",
       "      <td>CML277</td>\n",
       "      <td>1003</td>\n",
       "      <td>Zm00024ab225870</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>Zm00001d014121</td>\n",
       "      <td>CML322</td>\n",
       "      <td>1012</td>\n",
       "      <td>Zm00025ab231000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>Zm00001d014121</td>\n",
       "      <td>CML333</td>\n",
       "      <td>1005</td>\n",
       "      <td>Zm00026ab225150</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>Zm00001d014121</td>\n",
       "      <td>CML52</td>\n",
       "      <td>1011</td>\n",
       "      <td>Zm00019ab212360</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>Zm00001d014121</td>\n",
       "      <td>CML69</td>\n",
       "      <td>1005</td>\n",
       "      <td>Zm00020ab225370</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>Zm00001d014121</td>\n",
       "      <td>Col-0</td>\n",
       "      <td>334</td>\n",
       "      <td>AT1G11610</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>Zm00001d014121</td>\n",
       "      <td>HP301</td>\n",
       "      <td>1011</td>\n",
       "      <td>Zm00027ab226650</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>Zm00001d014121</td>\n",
       "      <td>Il14H</td>\n",
       "      <td>1005</td>\n",
       "      <td>Zm00028ab228830</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>Zm00001d014121</td>\n",
       "      <td>Ki11</td>\n",
       "      <td>1006</td>\n",
       "      <td>Zm00030ab223760</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>Zm00001d014121</td>\n",
       "      <td>Ki3</td>\n",
       "      <td>1012</td>\n",
       "      <td>Zm00029ab230310</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>Zm00001d014121</td>\n",
       "      <td>Ky21</td>\n",
       "      <td>1005</td>\n",
       "      <td>Zm00031ab231720</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>Zm00001d014121</td>\n",
       "      <td>M162W</td>\n",
       "      <td>1011</td>\n",
       "      <td>Zm00033ab235270</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>Zm00001d014121</td>\n",
       "      <td>M37W</td>\n",
       "      <td>1005</td>\n",
       "      <td>Zm00032ab229930</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>Zm00001d014121</td>\n",
       "      <td>Mo18W</td>\n",
       "      <td>1011</td>\n",
       "      <td>Zm00034ab236230</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>Zm00001d014121</td>\n",
       "      <td>Ms71</td>\n",
       "      <td>1005</td>\n",
       "      <td>Zm00035ab229720</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>Zm00001d014121</td>\n",
       "      <td>NC350</td>\n",
       "      <td>1011</td>\n",
       "      <td>Zm00036ab227950</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>Zm00001d014121</td>\n",
       "      <td>NC358</td>\n",
       "      <td>1011</td>\n",
       "      <td>Zm00037ab224700</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>Zm00001d014121</td>\n",
       "      <td>Oh43</td>\n",
       "      <td>1012</td>\n",
       "      <td>Zm00039ab227340</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>Zm00001d014121</td>\n",
       "      <td>Oh7B</td>\n",
       "      <td>1011</td>\n",
       "      <td>Zm00038ab226720</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>Zm00001d014121</td>\n",
       "      <td>Osativa</td>\n",
       "      <td>589</td>\n",
       "      <td>LOC_Os10g30390</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>Zm00001d014121</td>\n",
       "      <td>P39</td>\n",
       "      <td>1010</td>\n",
       "      <td>Zm00040ab237320</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>Zm00001d014121</td>\n",
       "      <td>Pvirgatum</td>\n",
       "      <td>804</td>\n",
       "      <td>Pavir.9KG201700</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>Zm00001d014121</td>\n",
       "      <td>Sbicolor</td>\n",
       "      <td>878</td>\n",
       "      <td>Sobic.001G235500</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>Zm00001d014121</td>\n",
       "      <td>Sitalica</td>\n",
       "      <td>776</td>\n",
       "      <td>Seita.9G241000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>Zm00001d014121</td>\n",
       "      <td>Sviridis</td>\n",
       "      <td>777</td>\n",
       "      <td>Sevir.9G241000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>Zm00001d014121</td>\n",
       "      <td>Tx303</td>\n",
       "      <td>1012</td>\n",
       "      <td>Zm00041ab231470</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>Zm00001d014121</td>\n",
       "      <td>Tzi8</td>\n",
       "      <td>1012</td>\n",
       "      <td>Zm00042ab231900</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>Zm00001d014121</td>\n",
       "      <td>Wm82</td>\n",
       "      <td>434</td>\n",
       "      <td>Glyma.07G160700</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        subject_id query_genotype  bit_score          query_id\n",
       "0   Zm00001d014121          B73v4       1015    Zm00001d014121\n",
       "1   Zm00001d014121          B73v5       1013   Zm00001eb222540\n",
       "2   Zm00001d014121            B97       1007   Zm00018ab231330\n",
       "3   Zm00001d014121         CML103       1012   Zm00021ab225420\n",
       "4   Zm00001d014121         CML228       1012   Zm00022ab224970\n",
       "5   Zm00001d014121         CML247       1005   Zm00023ab228620\n",
       "6   Zm00001d014121         CML277       1003   Zm00024ab225870\n",
       "7   Zm00001d014121         CML322       1012   Zm00025ab231000\n",
       "8   Zm00001d014121         CML333       1005   Zm00026ab225150\n",
       "9   Zm00001d014121          CML52       1011   Zm00019ab212360\n",
       "10  Zm00001d014121          CML69       1005   Zm00020ab225370\n",
       "11  Zm00001d014121          Col-0        334         AT1G11610\n",
       "12  Zm00001d014121          HP301       1011   Zm00027ab226650\n",
       "13  Zm00001d014121          Il14H       1005   Zm00028ab228830\n",
       "14  Zm00001d014121           Ki11       1006   Zm00030ab223760\n",
       "15  Zm00001d014121            Ki3       1012   Zm00029ab230310\n",
       "16  Zm00001d014121           Ky21       1005   Zm00031ab231720\n",
       "17  Zm00001d014121          M162W       1011   Zm00033ab235270\n",
       "18  Zm00001d014121           M37W       1005   Zm00032ab229930\n",
       "19  Zm00001d014121          Mo18W       1011   Zm00034ab236230\n",
       "20  Zm00001d014121           Ms71       1005   Zm00035ab229720\n",
       "21  Zm00001d014121          NC350       1011   Zm00036ab227950\n",
       "22  Zm00001d014121          NC358       1011   Zm00037ab224700\n",
       "23  Zm00001d014121           Oh43       1012   Zm00039ab227340\n",
       "24  Zm00001d014121           Oh7B       1011   Zm00038ab226720\n",
       "25  Zm00001d014121        Osativa        589    LOC_Os10g30390\n",
       "26  Zm00001d014121            P39       1010   Zm00040ab237320\n",
       "27  Zm00001d014121      Pvirgatum        804   Pavir.9KG201700\n",
       "28  Zm00001d014121       Sbicolor        878  Sobic.001G235500\n",
       "29  Zm00001d014121       Sitalica        776    Seita.9G241000\n",
       "30  Zm00001d014121       Sviridis        777    Sevir.9G241000\n",
       "31  Zm00001d014121          Tx303       1012   Zm00041ab231470\n",
       "32  Zm00001d014121           Tzi8       1012   Zm00042ab231900\n",
       "33  Zm00001d014121           Wm82        434   Glyma.07G160700"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# This is a stupid function but it seems to work correctly.\n",
    "df.sort_values(['bit_score'], ascending=False).groupby([\"subject_id\", \"query_genotype\"]).agg({\"bit_score\": \"first\", \"query_id\": \"first\",}).reset_index()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "subsequent-greek",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>subject_id</th>\n",
       "      <th>query_id</th>\n",
       "      <th>bit_score</th>\n",
       "      <th>subject_genotype</th>\n",
       "      <th>query_genotype</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>107</th>\n",
       "      <td>Zm00001d011673</td>\n",
       "      <td>Zm00018ab378850</td>\n",
       "      <td>709</td>\n",
       "      <td>B73v4</td>\n",
       "      <td>B97</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>108</th>\n",
       "      <td>Zm00001d011673</td>\n",
       "      <td>Zm00018ab160760</td>\n",
       "      <td>688</td>\n",
       "      <td>B73v4</td>\n",
       "      <td>B97</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>109</th>\n",
       "      <td>Zm00001d011673</td>\n",
       "      <td>Zm00018ab359540</td>\n",
       "      <td>615</td>\n",
       "      <td>B73v4</td>\n",
       "      <td>B97</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "         subject_id         query_id  bit_score subject_genotype  \\\n",
       "107  Zm00001d011673  Zm00018ab378850        709            B73v4   \n",
       "108  Zm00001d011673  Zm00018ab160760        688            B73v4   \n",
       "109  Zm00001d011673  Zm00018ab359540        615            B73v4   \n",
       "\n",
       "    query_genotype  \n",
       "107            B97  \n",
       "108            B97  \n",
       "109            B97  "
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df[df[\"query_genotype\"].isin([\"B97\", \"dsa\"])]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "deluxe-rendering",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['CML247',\n",
       " 'CML69',\n",
       " 'B73v5',\n",
       " 'CML333',\n",
       " 'Ki3',\n",
       " 'Sbicolor',\n",
       " 'CML277',\n",
       " 'Oh7B',\n",
       " 'CML52',\n",
       " 'Il14H',\n",
       " 'Sitalica',\n",
       " 'P39',\n",
       " 'M37W',\n",
       " 'Oh43',\n",
       " 'HP301',\n",
       " 'Tzi8',\n",
       " 'Osativa',\n",
       " 'Mo18W',\n",
       " 'Ki11',\n",
       " 'M162W',\n",
       " 'Col-0',\n",
       " 'B73v4',\n",
       " 'Ms71',\n",
       " 'Sviridis',\n",
       " 'Ky21',\n",
       " 'NC350',\n",
       " 'CML228',\n",
       " 'CML103',\n",
       " 'NC358',\n",
       " 'CML322',\n",
       " 'Tx303',\n",
       " 'Pvirgatum',\n",
       " 'B97',\n",
       " 'Wm82']"
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "test = pd.read_csv(os.path.join(\"init/BBHs_combs.tsv\"), sep=\"\\t\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "environmental-fireplace",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "test not in\n"
     ]
    }
   ],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
